---
title: "hiPSC models of Schizophrenia"
subtitle: 'Public release of data and code from Hoffman, et al. (submitted)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.Date()`"
documentclass: article
output: 
  html_document:
  toc: true
  smart: false
---

Data is available on Synapse.

## Summary
This includes code and plots for the final version of the paper.  Exploratory analysis and intermediate processing steps are too extensive to be included here


<!--- 
# run analysis
cd /home/ghoffman/workspace/COS_public_release
rmarkdown::render("analysis_public_release.Rmd", output_dir='./', intermediates_dir='./')
--->


```{r initialize, cache=FALSE, echo=FALSE, message=FALSE, results='hide'}
nthreads = 12
```

```{r load.always, cache=FALSE, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages(library(doParallel))
suppressPackageStartupMessages(library(synapseClient))

cl <- makeCluster(nthreads)
registerDoParallel(cl)

synapseLogin()

folder = c( "./results/differential_expression/Schizophrenia", 
            "./results/differential_expression/Cell_Composition/", 
            "./results/differential_expression/Sendai/", 
            "./results/variancePartition/",
            "./results/differential_expression/enrichment/",
            "./results/differential_expression/cellType/", 
            "./results/differential_expression/Schizophrenia/CTC_score/")

sapply(folder, function(x) {
  if( !file.exists(x)){
    dir.create(x, recursive=TRUE)
  }
  })
```

```{r load.packages, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages(library(limma))
suppressPackageStartupMessages(library(edgeR))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(variancePartition))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(ape))
suppressPackageStartupMessages(library(mixtools))
suppressPackageStartupMessages(library(latex2exp))
suppressPackageStartupMessages(library(readr))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(preprocessCore))
suppressPackageStartupMessages(library(colortools))
suppressPackageStartupMessages(library(xtable))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(qvalue))
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(GGally))
suppressPackageStartupMessages(library(gage))
suppressPackageStartupMessages(library(HTSanalyzeR))
suppressPackageStartupMessages(library(GSEABase))
suppressPackageStartupMessages(library(foreach))
suppressPackageStartupMessages(library(pheatmap))

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=TRUE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```


```{r ESS_ranges_cost}
# Define code for effective sample size calculations
plot_Ne = function(N_total, w){

  # total cost.  Equals total number of experiments when w 1
  # N_total = 100

  # given that the first experiment from a donor has cost 1,
  # w is the cost of each successive experiment from that same donor
  # w = .1

  rho = seq(0, 1, length.out=100)
  df = data.frame(m=c(), Ne=c(), rho=c())

  m_array = c(1:4)
  for(m in m_array){
    k = N_total / (1 + (m-1)*w)

    Ne  = (m*k) / (1+rho*(m-1))

    df = rbind(df, data.frame(m=as.character(m), Ne, rho))
  }

  ggplot(df, aes(rho, Ne, colour=m)) + geom_line(size=2) + theme_bw() +xlab("Donor effect (%)") + ylab("Effective sample size") +
    ggtitle(paste("Relative cost of second+ sample:", w)) + theme(plot.title=element_text(hjust=0.5))  + 
    scale_colour_manual(values=c("#CD2626", "#7ACD26", "#26CDCD", "#7A26CD"), name = "replicates\nper donor")
}

# plot_Ne(100, 1) + ylim(0, 160)
```

```{r get.MSigDB, echo=FALSE, results="hide"}
# define gene set enrichment analysis code
enrich_mSigDB = function( geneScoreEntrez, mSigDB, cutoff, minGeneSetSize=20){

  res = foreach( key=names(mSigDB), .packages=c("HTSanalyzeR") ) %dopar% {

    res = multiHyperGeoTest( collectionOfGeneSets=mSigDB[[key]], universe=names(geneScoreEntrez), hits=names(geneScoreEntrez)[geneScoreEntrez <= cutoff], minGeneSetSize=minGeneSetSize, verbose=FALSE)

    res = data.frame(res, stringsAsFactors=FALSE, check.names=FALSE)

    res$Class = rep(key, nrow(res))
    res[["Gene Set"]] = rownames(res)
    rownames(res) = c()
    res[["OR"]] = res[["Observed Hits"]] / res[["Expected Hits"]]
    res
  }

  res2 = c()
  for( i in 1:length(res)){
    res2 = rbind(res2, res[[i]])
  }

  res3 = data.frame(res2[,c("Class", "Gene Set", "Gene Set Size", "Expected Hits", "Observed Hits", "OR", "Pvalue")], stringsAsFactors=FALSE, check.names=FALSE)
  
  res3 = res3[order(res3$Pvalue),]

  return( res3 )
}


read.mSigDB = function(files, standard=TRUE){

    mSigDB = list()

    for( file in files){

      if( standard ){
          key = substring(basename(file), 1,2)
        }else{
          key = sapply(strsplit(basename(file), "\\."), function(x) x[1])
        }

        mSigDB[[key]] = readList( file )
    }
    return(mSigDB)
}

formTable = function(x,digits, stop=40){
  if(missing(digits)){
    digits = rep(2, ncol(x))
  }
  if( length(digits) != ncol(x)){
    stop("Lengths don't match")
  }
  x = data.frame(x)
  ret = sapply(1:ncol(x), function(i){
    format(x[,i,drop=FALSE],digits=digits[i])})
  ret = do.call("cbind", ret)
  rownames(ret) = rownames(x)
  ret[,1] = substring(ret[,1], 1, stop)
  rownames(ret) = substring(rownames(ret), 1, stop)
  ret
}


plot_within_btw = function( simMat, design, col=c("lightblue", "lightgreen"),main='',ylim=NULL,...){

  variation = list(type=c(), value=c(), donor1=c(), donor2=c())
  for(i in 1:ncol(design)){
    idx1 = which(design[,i] == 1)   
    idx0 = which(design[,i] == 0)
    C_sub = simMat[idx1, idx1]

    variation$type = append( variation$type, rep("Same Donor", sum(lower.tri(C_sub))) )  
    variation$value = append( variation$value, C_sub[lower.tri(C_sub)])   
    variation$donor1 = append(variation$donor1, rep(colnames(design)[i], length(C_sub[lower.tri(C_sub)]) ))      
    variation$donor2 = append(variation$donor2, rep(colnames(design)[i], length(C_sub[lower.tri(C_sub)]) ))

    variation$type = append( variation$type, rep("Different Donor", length(simMat[idx1,idx0])) )  
    variation$value = append( variation$value, simMat[idx1,idx0])    
    variation$donor1 = append(variation$donor1, rep("", length(simMat[idx1,idx0])) )      
    variation$donor2 = append(variation$donor2, rep("", length(simMat[idx1,idx0])))
  }
  variation = as.data.frame(variation)

  fig = ggplot( variation, aes(type, value)) + geom_violin(scale = "width", aes(fill=type)) + 
    ylab("Variance explained (%)") + xlab("") + geom_boxplot(width = 0.07, 
        fill = "grey", outlier.color = "black") + theme_bw(15) + 
        scale_fill_manual(values = col) + theme(legend.position = "none") + 
        ylab("Correlation between samples") + scale_fill_manual(values=c("grey", "steelblue1"))

    if( main !='' ){
      fig = fig + ggtitle( main )
    }
    if( !is.null(ylim) ){
      fig = fig + ylim( ylim )
    }

    fig = fig + theme(aspect.ratio=1)

    return(list(fig=fig, variation=variation))
}


# get files based on parent
geneset.files = synQuery('select * from file where parentId == "syn9908164"')

# Download files
files = sapply( geneset.files$file.id, function(file) synGet(file)@filePath)

geneSets = read.mSigDB(files, standard=FALSE )

geneSetsCombined_MSigDB = unlist( geneSets, recursive=FALSE)

i = grep("^c3", names(geneSetsCombined_MSigDB), invert=TRUE)
geneSetsCombined_MSigDB = geneSetsCombined_MSigDB[i]
i = grep("^c1", names(geneSetsCombined_MSigDB), invert=TRUE)
geneSetsCombined_MSigDB = geneSetsCombined_MSigDB[i]

# read magna genesets
geneset.files = synQuery('select * from file where parentId == "syn9908161"')

# Download files
files = sapply( geneset.files$file.id, function(file) synGet(file)@filePath)

geneSetsMagma = unlist(read.mSigDB(files, standard=FALSE), recursive=FALSE)
names(geneSetsMagma) = gsub("magma_gene_sets.", "", names(geneSetsMagma))


geneSetsCombined = c(geneSetsCombined_MSigDB, geneSetsMagma)

# remove small genesets
geneSetsCombined = geneSetsCombined[sapply(geneSetsCombined, length) > 20]
geneSetsCombined = geneSetsCombined[sapply(geneSetsCombined, length) < 5000]
# length(geneSetsCombined)
```

```{r romerParallel}
# Gabriel Hoffman
# April 3, 2017
#
# Parallelize romer when statistic="mean"

library(compiler)
library(foreach)
library(limma)
library(foreach)

##  ROMER.R
romerParallel <- function(y,...) UseMethod("romerParallel")

romerParallel.default <- cmpfun(function(y,index,design=NULL,contrast=ncol(design),array.weights=NULL,block=NULL,correlation=NULL,set.statistic="mean",nrot=9999,shrink.resid=TRUE,n_slices = 100,...)
# rotation mean-rank version of GSEA (gene set enrichment analysis) for linear models
# Gordon Smyth and Yifang Hu
# 27 March 2009.  Last modified 3 May 2015.
{
# Issue warning if extra arguments found
  dots <- names(list(...))
  if(length(dots)) warning("Extra arguments disregarded: ",sQuote(dots))

# Check y
  y <- as.matrix(y)
  ngenes <- nrow(y)
  n <- ncol(y)

# Check index
  if(!is.list(index)) index <- list(set=index)
  nsets <- length(index)
  if(nsets==0) stop("index is empty")
  SetSizes <- unlist(lapply(index,length))

# Check design
  if(is.null(design))
    stop("design matrix not specified")
  else {
    design <- as.matrix(design)
    if(mode(design) != "numeric") stop("design must be a numeric matrix")
  }
  if(nrow(design) != n) stop("row dimension of design matrix must match column dimension of data")
  ne <- nonEstimable(design)
  if(!is.null(ne)) cat("Coefficients not estimable:",paste(ne,collapse=" "),"\n")
  p <- ncol(design)
  if(p<2L) stop("design needs at least two columns")
  p0 <- p-1L
  d <- n-p

# Reform design matrix so that contrast of interest is last column
  if(length(contrast) == 1L) {
    contrast <- round(contrast)
    if(contrast < p) {
      i <- 1L:p
      design <- design[,c(i[-contrast],contrast)]
    }
  } else {
    design <- contrastAsCoef(design=design, contrast=contrast, first=FALSE)$design
  }

# Divide out array weights, if they exist
  if(!is.null(array.weights)) {
    if(any(array.weights <= 0)) stop("array.weights must be positive")
    if(length(array.weights) != n) stop("Length of array.weights doesn't match number of array")
    design <- design*sqrt(array.weights)
    y <- t(t(y)*sqrt(array.weights))
  }

# Divide out correlation structure, it is exists
  if(!is.null(block)) {
    block <- as.vector(block)
    if (length(block) != n) stop("Length of block does not match number of arrays")
    ub <- unique(block)
    nblocks <- length(ub)
    Z <- matrix(block,n,nblocks) == matrix(ub,n,nblocks,byrow = TRUE)
    cormatrix <- Z %*% (correlation * t(Z))
    diag(cormatrix) <- 1
    R <- chol(cormatrix)
    y <- t(backsolve(R, t(y), transpose = TRUE))
    design <- backsolve(R, design, transpose = TRUE)
  }

#   Fit linear model to all genes
  qr <- qr(design)
  signc <- sign(qr$qr[p,p])
  effects <- qr.qty(qr,t(y))

# Sample variances
  s2 <- colMeans(effects[-(1:p),,drop=FALSE]^2)

#   Estimate global hyper-parameters s0 and d0
  sv <- squeezeVar(s2,df=d)
  d0 <- sv$df.prior
  s02 <- sv$var.prior
  sd.post <- sqrt(sv$var.post)

# t-statistics and effects (orthogonal residuals)
  Y <- effects[-(1:p0),,drop=FALSE]
  YY <- colSums(Y^2)
  B <- Y[1,]
  modt <- signc*B/sd.post

  ##  EMP BAYES SHRINK FIRST EFFECT TO REMOVE SYSTEMATIC COMPONENT
  if(shrink.resid) {

  # Estimate hyperparameter p0 (proportion of DE genes)
    p.value <- 2*pt(abs(modt),df=d0+d,lower.tail=FALSE)
    proportion <- 1-propTrueNull(p.value) # proportion of DE probes

  # Estimate hyperparameter v0 (var.prior, variance of true logFC)
    stdev.unscaled <- rep_len(1/abs(qr$qr[qr$rank,qr$rank]),ngenes)
    var.unscaled <- stdev.unscaled^2
    df.total <- rep_len(d,ngenes) + sv$df.prior
    stdev.coef.lim <- c(0.1, 4)
    var.prior.lim <- stdev.coef.lim^2/sv$var.prior
    var.prior <- tmixture.vector(modt, stdev.unscaled, df.total, proportion, var.prior.lim)
    if (any(is.na(var.prior))) {
      var.prior[is.na(var.prior)] <- 1/sv$var.prior
      warning("Estimation of var.prior failed - set to default value")
    }

  # Estimate posterior probability of DE
    r <- (var.unscaled + var.prior)/var.unscaled
    if (sv$df.prior > 10^6)
      kernel <- modt^2 * (1 - 1/r)/2
    else
      kernel <- (1 + df.total)/2 * log((modt^2 + df.total)/(modt^2/r + df.total))
    lods <- log(proportion/(1 - proportion)) - log(r)/2 + kernel
    ProbDE <- exp(lods)/(1+exp(lods))

  # Shrink contrast to be like a residual
    Y[1,] <- Y[1,]*sqrt(var.unscaled/(var.unscaled+var.prior*ProbDE))

  }
##  END SHRINK

  set.statistic <- match.arg(set.statistic,c("mean","floormean","mean50"))

  if(set.statistic=="mean") {

    # Observed rankings for each set
    obs.ranks <- matrix(0,ngenes,3)
    obs.ranks[,1] <- rank(modt)
    obs.ranks[,2] <- ngenes-obs.ranks[,1]+1
    obs.ranks[,3] <- rank(abs(modt))

    AllIndices <- unlist(index)
    Set <- rep(1:nsets,SetSizes)
    obs.set.ranks <- rowsum(obs.ranks[AllIndices,],group=Set,reorder=FALSE)
    obs.set.ranks <- obs.set.ranks / SetSizes

    # Random rotations to simulate null hypothesis
    rot.ranks <- obs.ranks
    # p.value <- matrix(0,nrow=nsets,ncol=3)

    # n_slices = 100

    p.value = foreach( j = 1:n_slices, .combine=function(A, B){ A+B}, .packages="foreach") %dopar% {
      cat("Slice:", j, "\n")
      foreach(i=1:ceiling(nrot/n_slices), .combine=function(A, B){ A+B}) %do% {
        R <- matrix(rnorm((d+1)),1,d+1)
        R <- R/sqrt(rowSums(R^2))
        Br <- R %*% Y
        s2r <- (YY-Br^2)/d

        if(is.finite(d0)){
          sdr.post <- sqrt((d0*s02+d*s2r)/(d0+d))
        }else{
          sdr.post <- sqrt(s02)
        }

        modtr <- signc*Br/sdr.post
      
        rot.ranks[,1] <- rank(modtr)
        rot.ranks[,2] <- ngenes-rot.ranks[,1]+1
        rot.ranks[,3] <- rank(abs(modtr))

        rot.set.ranks <- rowsum(rot.ranks[AllIndices,],group=Set,reorder=FALSE)
        rot.set.ranks <- rot.set.ranks / SetSizes
        # p.value <- p.value + (rot.set.ranks >= obs.set.ranks)

        rm(R, Br, s2r, sdr.post, modtr)#, rot.ranks)
        (rot.set.ranks >= obs.set.ranks)
      }
    } 

  } # end "mean"

  if(set.statistic=="floormean") {

  # Observed rankings for each set
    obs.ranks <- matrix(0,ngenes,3)
    obs.ranks[,1] <- rank(pmax(modt,0))
    obs.ranks[,2] <- rank(pmax(-modt,0))
    obs.ranks[,3] <- rank(pmax(abs(modt),1))

    AllIndices <- unlist(index)
    Set <- rep(1:nsets,SetSizes)
    obs.set.ranks <- rowsum(obs.ranks[AllIndices,],group=Set,reorder=FALSE)
    obs.set.ranks <- obs.set.ranks / SetSizes

  # Random rotations to simulate null hypothesis
    rot.ranks <- obs.ranks
    p.value <- matrix(0,nrow=nsets,ncol=3)
    for(i in 1:nrot)
    {
      R <- matrix(rnorm((d+1)),1,d+1)
      R <- R/sqrt(rowSums(R^2))
      Br <- R %*% Y
      s2r <- (YY-Br^2)/d

      if(is.finite(d0))
        sdr.post <- sqrt((d0*s02+d*s2r)/(d0+d))
      else
        sdr.post <- sqrt(s02)

      modtr <- signc*Br/sdr.post
    
      rot.ranks[,1] <- rank(pmax(modtr,0))
      rot.ranks[,2] <- rank(pmax(-modtr,0))
      rot.ranks[,3] <- rank(pmax(abs(modtr),1))

      rot.set.ranks <- rowsum(rot.ranks[AllIndices,],group=Set,reorder=FALSE)
      rot.set.ranks <- rot.set.ranks / SetSizes
      p.value <- p.value + (rot.set.ranks >= obs.set.ranks)
    } 
  } # end "floormean"

  if(set.statistic=="mean50") {
  
  # Observed rankings for each set
    s.r <- rank(modt)
    s.abs.r <- rank(abs(modt))

    s.rank.mixed <- rep(0,nsets)
    s.rank.up <- rep(0,nsets)
    s.rank.down <- rep(0,nsets)

    m <- floor((SetSizes+1)/2)
    for(i in 1:nsets)
    {
      mh <- limma:::.meanHalf(s.r[index[[i]]],m[i])
      s.rank.up[i] <- mh[2] 
      s.rank.down[i] <- mh[1]
      s.rank.mixed[i] <- limma:::.meanHalf(s.abs.r[index[[i]]],m[i])[2]
    } 

  # Random rotations
    p.value <- matrix(rep(0,nsets*3),nrow=nsets,ncol=3)
    for(i in 1:nrot)
    {
      R <- matrix(rnorm((d+1)),1,d+1)
      R <- R/sqrt(rowSums(R^2))
      Br <- R %*% Y
      s2r <- (YY-Br^2)/d

      if(is.finite(d0))
        sdr.post <- sqrt((d0*s02+d*s2r)/(d0+d))
      else
        sdr.post <- sqrt(s02)

      modtr <- signc*Br/sdr.post

      s.r.2 <- rank(modtr)
      s.abs.r.2 <- rank(abs(modtr))
    
      for(j in 1:nsets)
      {
        mh.2 <- limma:::.meanHalf(s.r.2[index[[j]]],m[j])

        s.rank.up.2 <- mh.2[2]
        s.rank.down.2 <- mh.2[1]
        s.rank.mixed.2 <- limma:::.meanHalf(s.abs.r.2[index[[j]]],m[j])[2]
      
        if(s.rank.up.2 >= s.rank.up[j]) p.value[j,1] <- p.value[j,1]+1
        if(s.rank.down.2 <= s.rank.down[j]) p.value[j,2] <- p.value[j,2]+1
        if(s.rank.mixed.2 >= s.rank.mixed[j]) p.value[j,3] <- p.value[j,3]+1
      }
    } 
  } # end "mean50"

  p.value <- (p.value+1)/(nrot+1)
  colnames(p.value) <- c("Up","Down","Mixed")
  SetNames <- names(index)
  if(is.null(SetNames))
    rownames(p.value) <- 1:nsets
  else
    rownames(p.value) <- SetNames
  cbind(NGenes=SetSizes,p.value)
})
```


```{r load.data}
# get gene counts and expression cutoff
geneCounts = as.matrix(read.csv(getFileLocation(synGet( 'syn9908113' )), header=TRUE, row.names=1, check.names=FALSE))
# exonCounts = as.matrix(read.csv(getFileLocation(synGet( 'syn7113739' )), header=TRUE, row.names=1, check.names=FALSE))
isexpr = read.csv(getFileLocation(synGet( 'syn9908114' )), row.names=1, header=FALSE)[,1]
# isexpr_exon = read.csv(getFileLocation(synGet( 'syn7113733' )), row.names=1, header=FALSE)[,1]

# get gene annotations
geneInfo = read.table(getFileLocation(synGet( 'syn9909914' )), header=TRUE, stringsAsFactors=FALSE, sep="\t")

# get my metadata
info = read.csv(getFileLocation(synGet( 'syn9908174' )), header=TRUE)
info$Donor = factor(info$Donor)
idx = grep("(N|F)$", info$ID, invert=TRUE)
info$ID[idx] = gsub("-2$", "", as.character(info$ID[idx]))
colnames(geneCounts)[idx] = gsub("-2$", "R", as.character(colnames(geneCounts)[idx]))
# colnames(exonCounts)[idx] = gsub("-2$", "R", as.character(colnames(exonCounts)[idx]))

# get Brig's metdata
metadata = read.csv(getFileLocation(synGet( 'syn9908107' )), header=TRUE)
metadata$Cell.Type = factor(gsub(' ', '_',metadata$Cell.Type), c('6_wk_FB_neuron', 'NPC'))
# metadata = droplevels(metadata[-nrow(metadata),])
metadata$Donor = factor(sapply(strsplit(as.character(metadata$Sample.Name), '-'), function(x) x[1]))
metadata$Cell.Type_make.names = make.names(metadata$Cell.Type)
metadata$Exclusion = as.character(metadata$Exclusion)
# check that sample are present
# which(! (colnames(geneCounts) %in% metadata$Sample.Name))

# only keep samples present in data and metadata
idx = match(colnames(geneCounts), metadata$Sample.Name)
metadata = metadata[idx,]

# add sendai reads
sendaiReads = read.csv(getFileLocation(synGet( 'syn9908176' )), header=TRUE, stringsAsFactors=FALSE)
sendaiReads$RNA_Sample_ID = gsub("-2$", "R",sendaiReads$RNA_Sample_ID)

idx = match(metadata$Sample.Name, sendaiReads$RNA_Sample_ID)
metadata$SendaiReads = sendaiReads$SendaiCounts[idx]

# Sendai results
metadata$totalReads = colSums(geneCounts)
metadata$SendaiCPM = with(metadata, (SendaiReads) / (totalReads/1e6))

# change mislabeling
i = metadata$Sample.Name %in% c("499-X-CF", "499-X-CN")
metadata$Donor = as.character(metadata$Donor)
metadata$Donor[i] = "449"
metadata$Sex[i] = "Female"
metadata$Donor = factor(metadata$Donor)
```

# Sendai Expression
```{r Figure_Sendai_expression, fig.width=4, fig.height=8}
i = order(metadata$SendaiCPM)
ggplot(metadata[i,], aes(reorder(Sample.Name, SendaiCPM), SendaiCPM)) + 
  theme_bw(8) + coord_flip() + geom_point() + xlab('') + ylab(expression("Sendai Expression" ~ (CPM))) +
  scale_y_log10() + theme(legend.position="none") 
```

# Biotype of expressed genes
```{r biotype.counts, results='asis'}
biotype.counts = sort(table(geneInfo$transcript_biotype[isexpr]), decreasing=TRUE)
df = data.frame(biotype.counts)
colnames(df) = c("Biotype", 'Count')

print(xtable( df ), include.rownames=FALSE )
```

# Check sex on 94 samples
```{r sexcheck.initial}
genes = DGEList(counts=geneCounts[isexpr,])
genes = calcNormFactors(genes)

geneExpr = rpkm( genes, geneInfo$Length[isexpr], log=FALSE )

ensGenes_XIST = with(geneInfo, Geneid[hgnc_symbol %in% c("XIST")])

genes_on_chrY = c( "USP9Y", "UTY", "NLGN4Y", 'ZFY', 'RPS4Y1', 'TXLNG2P')
ensGenes_chrY = with(geneInfo, Geneid[hgnc_symbol %in% genes_on_chrY])

data = data.frame( XIST = log2(geneExpr[ensGenes_XIST,]), genesY = log2(colSums(geneExpr[ensGenes_chrY,,drop=FALSE])),
   Sex = metadata$Sex, ID = metadata$Sample.Name)
data$Sex = as.character(data$Sex)

i = with(data, (XIST <  0 & genesY < 5) | (XIST > 2 & genesY > 4))
data$Sex[i] = "Female - problem"
data$Sex = factor(data$Sex, c("Male", "Female", "Female - problem" ))

ggplot(data, aes(XIST, genesY)) + geom_point(aes(colour=Sex)) + theme_bw(15) + 
    scale_colour_manual(values=c("blue", "red", "grey")) + theme(aspect.ratio=1) + 
    xlab(expression("XIST" ~ (log[2] ~ RPKM))) + ylab(expression("Sum of 6 chrY genes" ~ (log[2] ~ RPKM))) +
    geom_text_repel(data=data[i,], aes(XIST, genesY, label=ID))

kable(data.frame(genes_on_chrY))
```

# Detect sample contamination with VerifyBamID
The labeling of Donor 499 was corrected manually
```{r verifyBAMID, echo=FALSE, message=FALSE}
verifyBAM_RNA = read.csv(getFileLocation(synGet( 'syn9908178' )), header=TRUE, stringsAsFactors=FALSE, comment.char='')

data = melt(verifyBAM_RNA)
data$variable = as.character(data$variable)
data = data[-grep("FREEMIX", data$variable),]

data$variable[data$variable == 'CHIPMIX_PsychChip'] = "PsychChip"
data$variable[data$variable == 'CHIPMIX_WES'] = "Whole Exome Sequencing"
data$variable = factor(data$variable)

data$x = as.numeric(data$variable) + runif(nrow(data), -.2, .2)

ggplot(data, aes(variable, value) ) + geom_blank() + geom_point(aes(x, value)) + theme_bw(15) + ylab("Estimated % Contamination") + xlab("") + geom_text_repel(data=subset(data,value > .1), aes(x, value, label=RNA_Sample_ID), box.padding = unit(.4, 'lines')) +
  theme(axis.text.x=element_text(angle=30, hjust=1), aspect.ratio=1)
```


```{r exclude.samples, echo=FALSE, message=FALSE}
# More exclusions
excludeBasedOnSexGenes = c("1275-B-3F", "1275-B-3N", "1275-C-1F", "1275-C-1N", "2476-1-4F", "2476-1-4N", "2484-2aF", "2484-2aN", "3113-3-21F", "3113-3-21N", "3121-2-1F", "3121-2-1N", "3121-2-MSSM2F", "3121-2-MSSM2N", "676-1-2N")

metadata$Exclusion[metadata$Sample.Name %in% excludeBasedOnSexGenes] = 'yes' #"maybe"

# apply exclusion
idx_exclude = which(metadata$Exclusion == "yes")
```


## Excluded due to sample mislabeling, contamination or issues with expression on the sex chromosomes
```{r write.exclude, results="asis"}
print(xtable(metadata[idx_exclude,1,drop=FALSE]), include.rownames=FALSE)     
```

```{r resume.processing}
metadata = metadata[-idx_exclude,]   
geneCounts = geneCounts[,-idx_exclude]
# exonCounts = exonCounts[,-idx_exclude]  
metadata$color = rainbow(nlevels(metadata$Donor))[metadata$Donor]

# check if geneCounts and metadata are sorted correctly
# table(colnames(geneCounts) == as.character(metadata$Sample.Name))
```

#### Samples included in analysis
```{r dataset.info, results="asis"}
print(xtable(xtabs(~ Dx + Cell.Type, metadata)))
```

```{r dataset.case.control, results="asis"}
df = with(metadata, data.frame(Case = length(unique(Donor[Dx == 'SZ'])), Control =length(unique(Donor[Dx == 'CT']))))
print(xtable(df), include.rownames=FALSE)
```

# QC metrics
```{r SI_Fig_1}
qc_metrics = read.csv(getFileLocation(synGet( 'syn9908127' )), header=TRUE, stringsAsFactors=FALSE)

hist(qc_metrics$rRNA.rate, xlab="rRNA rate", main='rRNA rate')
```


```{r voom}
genes = DGEList(counts=geneCounts[isexpr,])
genes = calcNormFactors(genes)
design = model.matrix(~Cell.Type + Sex, metadata)

# voom / dupcor 1
#################

vobj_init = voom(genes, design, plot=FALSE )

dupcor_init = duplicateCorrelation( vobj_init, design, block=metadata$Donor)
# dupcor_init$consensus

# # voom / dupcor 2
# #################
# # re-estimate weights using correlation values

# dupcor = list(consensus = 0.56)
# dupcor_init = list(consensus = 0.56)
vobj <- voom(genes, design, block = metadata$Donor, correlation=dupcor_init$consensus)

dupcor = duplicateCorrelation( vobj, design, block=metadata$Donor)
# dupcor$consensus
```

```{r rpkm}
geneExprlog2RPKM = rpkm( genes, geneInfo$Length[isexpr], log=TRUE )
```

# Check sex on cleaned dataset
```{r sexcheck}
genes = DGEList(counts=geneCounts[isexpr,])
genes = calcNormFactors(genes)

geneExpr = rpkm( genes, geneInfo$Length[isexpr], log=FALSE )
ensGenes_XIST = with(geneInfo, Geneid[hgnc_symbol %in% c("XIST")])

genes_on_chrY = c( "USP9Y", "UTY", "NLGN4Y", 'ZFY', 'RPS4Y1', 'TXLNG2P')
ensGenes_chrY = with(geneInfo, Geneid[hgnc_symbol %in% genes_on_chrY])

data = data.frame( XIST = log2(geneExpr[ensGenes_XIST,]), genesY = log2(colSums(geneExpr[ensGenes_chrY,,drop=FALSE])),
   Sex = metadata$Sex, ID = metadata$Sample.Name)
data$Sex = as.character(data$Sex)

i = with(data, (XIST <  0 & genesY < 5) | (XIST > 2 & genesY > 4))
data$Sex[i] = "Female - problem"
data$Sex = factor(data$Sex, c("Male", "Female", "Female - problem" ))

ggplot(data, aes(XIST, genesY)) + geom_point(aes(colour=Sex)) + theme_bw(15) + 
    scale_colour_manual(values=c("blue", "red", "grey")) + theme(aspect.ratio=1) + 
    xlab(expression("XIST" ~ (log[2] ~ RPKM))) + ylab(expression("Sum of 6 chrY genes" ~ (log[2] ~ RPKM))) +
    geom_text_repel(data=data[i,], aes(XIST, genesY, label=ID))
```

# Differential gene expression based on Sendai virus exression
```{r Sendai.DE}

metadata$SendaiCPM = with(metadata, log2((SendaiReads+1) / (totalReads/1e6)))
metadata$SendaiCPM[is.na(metadata$SendaiCPM)] = -4.7

# rank normalize the Sendai virus CPM
a = rank(metadata$SendaiCPM)
metadata$SendaiCPM_norm = (a - mean(a)) / sd(a)

design = model.matrix(~ SendaiCPM_norm + Sex + Cell.Type + RIN, metadata)

fitNorm = lmFit( vobj, design, block=metadata$Donor, correlation=dupcor$consensus )
fitNorm = eBayes(fitNorm)

resDENormSendai = topTable(fitNorm, coef='SendaiCPM_norm', number=1e7 , sort.by='P')
resDENormSendai$qvalue = qvalue( resDENormSendai$P.Value )$qvalue
resDENormSendai$gene = geneInfo$geneName[match(rownames(resDENormSendai), geneInfo$Geneid)]
resDENormSendai$symbol = sub("^(ENSG.*)$", NA, resDENormSendai$gene)

file = "results/differential_expression/Sendai/COS_Sendai_DE.tsv"
df = data.frame(geneName = rownames(resDENormSendai), resDENormSendai)
write.table(df, file, quote=FALSE, row.names=FALSE, sep="\t")

# gene enrichment
values = resDENormSendai$qvalue 
names(values) = resDENormSendai$symbol
values = values[!is.na(names(values))]

resEnrichSendaiNorm = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = 0.05)

file = "results/differential_expression/Sendai/COS_enrichments_Sendai_DE.tsv"
write.table(resEnrichSendaiNorm[,-1], file, quote=FALSE, sep="\t", row.names=FALSE)
```

```{r, results="asis"}
normal_print(paste0("DE genes at FDR < 5%: ", sum(resDENormSendai$qvalue < .05)))
```

# Yaminaka factors
#### POU5F1 is not expressed at a sufficient level
```{r, yaminaka.de}
kable(resDENormSendai[resDENormSendai$gene %in% c("MYC", "SOX2", "KLF4", "POU5F1"),c(1:4, 7, 8)])
```

```{r, yaminaka.de2, fig.width=9, fig.height=3}
grid.table(resDENormSendai[resDENormSendai$gene %in% c("MYC", "SOX2", "KLF4", "POU5F1"),c(1:4, 7, 8)])
```

```{r, Sendai.DE.barplot, fig.height=12}
idx = grep("^c4", resEnrichSendaiNorm$`Gene Set`, invert=TRUE)
res = resEnrichSendaiNorm[idx,][1:30,]

res$`Gene Set` = factor( res$`Gene Set`, rev(res$`Gene Set`))

ggplot(res, aes(x=`Gene Set`, y=-log10(Pvalue))) + geom_bar(stat='identity', fill="navy") + 
  coord_flip() + theme_bw(10) + ylab(bquote(Enrichment~(-log[10]~p))) + xlab('')
```


# Compare Neurons and NPC
```{r DE.NPC_neuron}

form = ~ Dx + Cell.Type + Sex 

design = model.matrix( form, metadata)
fit = lmFit( vobj, design, block=metadata$Donor, correlation=dupcor$consensus )
fit = eBayes(fit)

resDE_cellType = topTable(fit, coef='Cell.TypeNPC', number=1e7 , sort.by='P')
resDE_cellType$qvalue = qvalue( resDE_cellType$P.Value )$qvalue
resDE_cellType$gene = geneInfo$geneName[match(rownames(resDE_cellType), geneInfo$Geneid)]
resDE_cellType$symbol = sub("^(ENSG.*)$", NA, resDE_cellType$gene)

file = "results/differential_expression/cellType/DE_NPC_neuron.tsv"
df = data.frame(geneName = rownames(resDE_cellType), resDE_cellType)
write.table( df, file=file, row.names=FALSE, quote=FALSE, sep="\t")

gene = "ENSG00000252196"
GE = data.frame(expr = vobj$E[gene,], cellType = metadata$Cell.Type)

# negative logFC is down in NPC
plotStratify( expr ~ cellType, GE, main=gene, ylab="Expression (log2 CPM)") + scale_fill_manual(values = c("orange", 'green2'))  + scale_color_manual(values = c("orange", 'green2'))


# Enrichment analysis
values = with(resDE_cellType, ifelse(qvalue < 0.05 & abs(logFC) > 1, 0, 1))
names(values) = resDE_cellType$symbol
values = values[!is.na(names(values))]
resEnrich = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = .05)

file = "results/differential_expression/cellType/DE_NPC_neuron_enrichments.tsv"
write.table( resEnrich, file=file, row.names=TRUE, quote=FALSE, sep="\t")


# Enrichment analysis: up in Neuron
values = with(resDE_cellType, ifelse(qvalue < 0.05 & logFC < -1, 0, 1))
names(values) = resDE_cellType$symbol
values = values[!is.na(names(values))]
resEnrich_neuron = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = .05)

file = "results/differential_expression/cellType/DE_enrichments_up_in_neuron.tsv"
write.table( resEnrich_neuron, file=file, row.names=TRUE, quote=FALSE, sep="\t")


# Enrichment analysis: up in NPC
values = with(resDE_cellType, ifelse(qvalue < 0.05 & logFC > 1, 0, 1))
names(values) = resDE_cellType$symbol
values = values[!is.na(names(values))]
resEnrich_NPC = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = .05)

file = "results/differential_expression/cellType/DE_enrichments_up_in_neuron.tsv"
write.table( resEnrich_NPC, file=file, row.names=TRUE, quote=FALSE, sep="\t")
```


```{r DE.NPC_neuron.volcano }
# Volcano plot
cutoff = 0.05
cutoff2 = 0.01
df = resDE_cellType
col = rev(c("n.s."="grey60", "< 5%"="firebrick", "< 1%"="red"))
df$color = names(col)[3]
df$color[df$qvalue < cutoff] = names(col)[2]
df$color[df$qvalue < cutoff2] = names(col)[1]
df$color = factor(df$color, names(col))
maxlogFC = max(abs(range(df$logFC)))
figVolcano = ggplot(df, aes(logFC, -log10(P.Value), colour=color)) + geom_point() + theme_bw(15) +
  ylab(bquote(-log[10]~p-value)) + xlab(bquote(log[2]~fold~change)) + 
  scale_color_manual(values = array(col), name="FDR") + theme(legend.position=c(.9,.9), aspect.ratio=1) + 
  xlim(-maxlogFC, maxlogFC)

  figVolcano
```

# Enrichments for genes up in Neurons
```{r DE.NPC_Neuron.enrichment.up_in_Neuron, results="asis"}
print(xtable( resEnrich_neuron[1:20,-1], digits=c(0, 0, 2, 2, 0, 2, -2) ), include.rownames=FALSE )
```

```{r, DE.NPC_Neuron.enrichment.up_in_Neuron_barplot, fig.height=12}
idx = grep("^c4", resEnrich_neuron$`Gene Set`, invert=TRUE)
res = resEnrich_neuron[idx,][1:30,]

res$`Gene Set` = factor( res$`Gene Set`, rev(res$`Gene Set`))

ggplot(res, aes(x=`Gene Set`, y=-log10(Pvalue))) + geom_bar(stat='identity', fill="orange") + 
  coord_flip() + theme_bw(10) + ylab(bquote(Enrichment~(-log[10]~p))) + xlab('')
```


# Enrichments for genes up in NPCs
```{r DE.NPC_Neuron.enrichment.up_in_NPC, results="asis"}
print(xtable( resEnrich_NPC[1:20,-1], digits=c(0, 0, 0, 2, 0, 2, -2) ), include.rownames=FALSE )
```


```{r, DE.NPC_Neuron.enrichment.up_in_NPC_barplot, fig.height=12}
idx = grep("^c4", resEnrich_NPC$`Gene Set`, invert=TRUE)
res = resEnrich_NPC[idx,][1:30,]

res$`Gene Set` = factor( res$`Gene Set`, rev(res$`Gene Set`))

ggplot(res, aes(x=`Gene Set`, y=-log10(Pvalue))) + geom_bar(stat='identity', fill="green2") + 
  coord_flip() + theme_bw(10) + ylab(bquote(Enrichment~(-log[10]~p))) + xlab('')
```




# Overlap between NPC and neuron
```{r NPC_neuron_overlap, fig.width=4, fig.height=4}
library(VennDiagram)
library(sets)

# figList = foreach(cutoff = 1:6) %do% {
jvalue = c()
i = 1
for(cutoff in 0:6)  {

  idx1 = metadata$Cell.Type == "6_wk_FB_neuron"
  idx2 = metadata$Cell.Type == "NPC"

  df = data.frame( neuron = rowMeans(geneExprlog2RPKM[,idx1]) > cutoff,
                   NPC    = rowMeans(geneExprlog2RPKM[,idx2]) > cutoff)

  df$both = with(df, neuron & NPC)

  jvalue[i] = set_similarity(as.set(rownames(df)[df$neuron]), as.set(rownames(df)[df$NPC]), method = "Jaccard")

  knit_print(paste("cutoff: log2 RPKM >", cutoff) )
  grid.newpage()
  draw.pairwise.venn(area1 = sum(df$neuron), area2 = sum(df$NPC), cross.area = sum(df$both), category = c("Neuron", 
      "NPC"), fill=c("orange", "green2"), lty = "blank", main="asdfasdf")
  i = i + 1
}

df = data.frame(cutoff = 0:6, Jaccard = jvalue)
```

```{r cell_type_jaccard}
ggplot(df, aes(cutoff, Jaccard)) + geom_bar(stat='identity') + theme_bw(14) + ylab("Jaccard similarity") + 
  xlab("log2 RPKM cutoff") + ylim(0, 1) + ggtitle("Set overlap between genes expressed in neurons and NPCs")
```




# Cibersort results
```{r cibersort}
cibersort = read.table(getFileLocation(synGet( 'syn9908124' )), row.names=1, header=TRUE)
rownames(cibersort) = gsub("-2$", "R", rownames(cibersort))

# missing samples
#  metadata$Sample.Name[!metadata$Sample.Name %in% rownames(cibersort)]

metadata2 = merge(metadata, cibersort, by.x='Sample.Name', by.y='row.names')
rownames(metadata2) = metadata2$Sample.Name
```

# Histogram of cell type composition: all samples
```{r cibersort.boxplot, message=FALSE}
ggplot(melt(cibersort*100), aes(reorder(variable, value, median), value, fill=variable)) + geom_boxplot() + theme_bw(15) + 
  theme(legend.position="none") + ylab("Cell composition score") + xlab("") + coord_flip()
```

## There is no case-control difference in cell type scores
```{r cibersort.disease, fig.width=6, fig.height=3.2}
resdf = matrix(ncol=3, nrow=0)
dfAll_diagnosis_CTC = matrix(ncol=3, nrow=0)

for(cellType in levels(metadata$Cell.Type)){
  for(score in colnames(cibersort)){

    form = paste(score, '~ Dx')

    idx = (metadata$Cell.Type == cellType)

    # # linear regression test
    # fit = lm( form, metadata2[idx,])
    # res = coef(summary(fit))[2,4]

    df = data.frame(Dx = metadata2$Dx[idx], score = metadata2[[score]][idx])

    res = wilcox.test( with(df, score[Dx=='CT']), with(df, score[Dx=='SZ']) )$p.value

    resdf = rbind(resdf, c(cellType, score, res))

    dfAll_diagnosis_CTC = rbind(dfAll_diagnosis_CTC, data.frame(cellType, Dx = df$Dx, CTC = df$score, score = score))
  }
}

colnames(resdf) = c("CellType", "score", 'pvalue')
resdf = data.frame(resdf, stringsAsFactors=FALSE)
resdf$pvalue = as.numeric(resdf$pvalue)

ggplot(resdf, aes(score, -log10(pvalue), fill=CellType)) + geom_bar(stat='identity', position=position_dodge()) + 
  theme_bw(15) + coord_flip() + scale_fill_manual(values=c("orange", "green2")) + xlab('') +
  geom_hline(yintercept=-log10(0.05/nrow(resdf)), lty=2) +
  geom_hline(yintercept=-log10(0.05), lty=3) +
  ylab(bquote(log[10]~p-value)) + theme(legend.position="none")
```

```{r plot_all_CTC_scores, fig.width=12, fig.height=16}

dfAll_diagnosis_CTC$cellType = as.character(dfAll_diagnosis_CTC$cellType)
dfAll_diagnosis_CTC$cellType[dfAll_diagnosis_CTC$cellType=="6_wk_FB_neuron"] = "iPSC-neuron"
dfAll_diagnosis_CTC$cellType[dfAll_diagnosis_CTC$cellType=="NPC"] = "iPSC-NPC"

ggplot(dfAll_diagnosis_CTC, aes(Dx, CTC)) + geom_boxplot(aes(fill=Dx)) + 
  facet_wrap(~ score + cellType, scales = "free_y", ncol=4) + xlab('') + 
  scale_fill_manual(values=setColors("deepskyblue2", 2), name="Diagnosis") + ylab("Cell Type Composition score")
```

# Cell composition in Neurons and NPC
```{r cibersort.stratify, fig.width=21, fig.height=28}
figList = list()
for( cell in colnames(cibersort)[order(tolower(colnames(cibersort)), decreasing=FALSE)] ){

  fit = wilcox.test( metadata2[[cell]][metadata2$Cell.Type == levels(metadata2$Cell.Type)[1]],
                     metadata2[[cell]][metadata2$Cell.Type == levels(metadata2$Cell.Type)[2]])

  fig = plotStratifyBy( metadata2, 'Cell.Type', cell, sort=FALSE, legend=FALSE, main=cell, xlab="Cell type", ylab="Composition score", text=paste("p =", format(fit$p.value, digits=2))) + scale_fill_manual(values=c("orange", "green2")) + scale_colour_manual(values=c("orange", "green2")) + ylim(0, max(metadata2[[cell]])) + theme(aspect.ratio=1)

  figList[[cell]] = fig + theme_bw(15) + theme(legend.position="none") + theme(plot.title=element_text(hjust=0.5))
}

do.call("grid.arrange", figList)
```


# Differential expression based on cell fraction
**Enrichment test for gene positively correlated with cell fraction**
```{r DE.cellFraction}

resDEComponments = list()
resEnrichComponents = list()
for(cellType in levels(metadata2$Cell.Type)){
  for(cellComponent in colnames(cibersort) ){
    
    key = paste(cellType, cellComponent, sep='-')

    idx = (metadata2$Cell.Type == cellType)

    # differential expression
    design = model.matrix( as.formula(paste('~ ', cellComponent)), metadata2[idx,]) # Cell.Type +

    if(var(design[,2]) < 1e-5){
      next
    }

    fit = lmFit( vobj[,idx], design, block=metadata$Donor[idx], correlation=dupcor$consensus )
    fit = eBayes(fit)

    # qvalue and prepare for enrichment
    resDEComponments[[key]] = topTable(fit, coef=cellComponent, number=1e7 , sort.by='P')
    resDEComponments[[key]]$qvalue = qvalue( resDEComponments[[key]]$P.Value )$qvalue
    resDEComponments[[key]]$gene = geneInfo$geneName[match(rownames(resDEComponments[[key]]), geneInfo$Geneid)]
    resDEComponments[[key]]$symbol = sub("^(ENSG.*)$", NA, resDEComponments[[key]]$gene)

    # enrichment analysis
    values = resDEComponments[[key]]$qvalue * sign(resDEComponments[[key]]$logFC)*(-1)
    names(values) = resDEComponments[[key]]$symbol
    values = values[!is.na(names(values))]
    resEnrichComponents[[key]] = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = .05)


    df = data.frame(geneName = rownames(resDEComponments[[key]]), resDEComponments[[key]])

    file = paste0("results/differential_expression/Cell_Composition/Cell_Composition_Score_DE_", key, ".tsv")
    write.table(df, file=file, quote=FALSE, row.names=FALSE, sep="\t")
  }
}
```

```{r DE.cellFraction.print, results='asis'}
for(key in names(resEnrichComponents)){
  # knit_print( key )
  print(kable(formTable(resEnrichComponents[[key]][1:10,-1]), caption=key, row.names=FALSE))

  file = paste0("results/differential_expression/Cell_Composition/Cell_Composition_Score_enrichments_", key, ".tsv")
  write.table(resEnrichComponents[[key]][,-1], file=file, quote=FALSE, row.names=FALSE, sep="\t")
}
```



# Analysis of original data
```{r PCA.original, fig.height=5}
dcmpOriginal = prcomp( t(vobj$E) )

# clustering
sampleDists <- dist( t( vobj$E ) )
sampleDistMatrix <- as.matrix( sampleDists )

hcl = hclust( sampleDists )

col = c("orange", "green2")
df = data.frame(dcmpOriginal$x, metadata2)

frac = dcmpOriginal$sdev^2 / sum(dcmpOriginal$sdev^2)   
xlab = paste0("PC1 (", format(100*frac[1], digits=3), "%)", sep='')
ylab = paste0("PC2 (", format(100*frac[2], digits=3), "%)", sep='')

ggplot(df, aes(PC1, PC2, color=Cell.Type)) + geom_point() + theme_bw(15) + xlab(xlab) + ylab(ylab) + 
  scale_color_manual(values=col) + theme(aspect.ratio=1)

```

```{r hclust.original, echo=FALSE, message=FALSE}
plot(as.phylo(hcl), type="fan", tip.color=rainbow(nlevels(metadata2$Donor))[metadata2$Donor], cex=.8)
```


```{r within_between.original, echo=FALSE, message=FALSE, fig.height=5}   
simList = list()
y_range = c()
for( cellType in rev(levels( metadata2$Cell.Type)) ){

  idx = (metadata2$Cell.Type == cellType) & ( metadata2$Exclusion != "maybe")

  C = cor( vobj$E, method="pearson"  )^2

  simList[[cellType]] = plot_within_btw( C[idx,idx], model.matrix(~ Donor-1, metadata2[idx,]), main=cellType) 
  y_range = c(y_range, min(C[idx,idx]))
}

do.call("grid.arrange", c(lapply(simList, function(x){

  pv = with(x$variation, wilcox.test(value[type == "Same Donor"], value[type == "Different Donor"], alternative="greater"))

  x$fig + theme(plot.title=element_text(hjust=0.5)) + ylim(min(y_range), 1)  + annotate("text", 2, .997, label=paste0("p < ", format(pv$p.value, digits=3)), size=6)
  }   
  ), ncol=2))
```

```{r compute.residuals}

metadata2$Myocyte_Neuron = with(metadata2, Myocyte*(gsub(" ", "_",Cell.Type)=="6_wk_FB_neuron"))
metadata2$Astrocytes_Neuron = with(metadata2, Astrocytes*(gsub(" ", "_",Cell.Type)=="6_wk_FB_neuron"))
metadata2$Neuron_Neuron = with(metadata2, Neuron*(gsub(" ", "_",Cell.Type)=="6_wk_FB_neuron"))
metadata2$MEF_Neuron = with(metadata2, MEF*(gsub(" ", "_",Cell.Type)=="6_wk_FB_neuron"))
metadata2$Fibroblast_Neuron = with(metadata2, Fibroblast*(gsub(" ", "_",Cell.Type)=="6_wk_FB_neuron"))

# form = ~ 0 + Cell.Type + Neuron:Cell.Type + Fibroblast_Neuron 
# form = ~ 0 + Cell.Type + Neuron:Cell.Type + Myocyte_Neuron + Astrocytes_Neuron + Sex
form = ~ 0 + Cell.Type + Fibroblast:Cell.Type + MEF:Cell.Type

fitList = fitVarPartModel( vobj, form, metadata2, fxn=function(fit){
  residuals(fit) + t(model.matrix(~ 0 + fit$model$Cell.Type) %*% coef(fit)[1:2])
  } )

exprResiduals = do.call(rbind, fitList)
colnames(exprResiduals) = metadata2$Sample.Name

# PCA
dcmpResiduals = prcomp( t(exprResiduals) )

# clustering
sampleDists <- dist( t( exprResiduals ) )
sampleDistMatrix <- as.matrix( sampleDists )

hcl = hclust( sampleDists )
```


# Analysis after correcting for cell type composition scores for fibroblasts and MEF
```{r PCA.residuals, cache=TRUE, fig.height=5}
frac = dcmpResiduals$sdev^2 / sum(dcmpResiduals$sdev^2)   
xlab = paste0("PC1 (", format(100*frac[1], digits=3), "%)", sep='')
ylab = paste0("PC2 (", format(100*frac[2], digits=3), "%)", sep='') 

df = data.frame(dcmpResiduals$x[,1:2], name=rownames(dcmpResiduals$x), cellType = metadata$Cell.Type)

ggplot(df, aes(PC1, PC2, color=cellType)) + geom_point() + theme_bw(15) + xlab(xlab) + ylab(ylab) + 
  scale_color_manual(values=col) + theme(aspect.ratio=1)
```

```{r hclust.residuals, cache=FALSE}
plot(as.phylo(hcl), type="fan", tip.color=rainbow(nlevels(metadata$Donor))[metadata$Donor], cex=.8)
```

```{r within_between.residuals, echo=FALSE, message=FALSE, cache=TRUE, fig.height=5}     
simList = list()
y_range = c() 
for( cellType in rev(levels( metadata2$Cell.Type)) ){

  idx = (metadata2$Cell.Type == cellType) & ( metadata2$Exclusion != "maybe")

  C = cor( exprResiduals, method="pearson"  )^2
  # C = cor( vobj$E, method="pearson" )

  simList[[cellType]] = plot_within_btw( C[idx,idx], model.matrix(~ Donor-1, metadata2[idx,]), main=cellType) 
  y_range = c(y_range, min(C[idx,idx]))
}

do.call("grid.arrange", c(lapply(simList, function(x){

  pv = with(x$variation, wilcox.test(value[type == "Same Donor"], value[type == "Different Donor"], alternative="greater"))

  x$fig + theme(plot.title=element_text(hjust=0.5)) + ylim(min(y_range), 1)  + annotate("text", 2, .997, label=paste0("p < ", format(pv$p.value, digits=3)), size=6)
  }   
  ), ncol=2))

```


# Link PCA to cell fraction
```{r PCA.cell.fraction}
 
dcmpAll = prcomp( t(vobj$E) )

idxNeuron = metadata2$Cell.Type == levels(metadata2$Cell.Type)[1]
idxNPC = metadata2$Cell.Type == levels(metadata2$Cell.Type)[2]

dcmpNeuron = prcomp( t(vobj$E[,idxNeuron]) )
dcmpNPC = prcomp( t(vobj$E[,idxNPC]) )

col = c("orange", "green2")

frac = dcmpAll$sdev^2 / sum(dcmpAll$sdev^2)
xlab = paste0("PC1 (", format(100*frac[1], digits=3), "%)", sep='')
ylab = paste0("PC2 (", format(100*frac[2], digits=3), "%)", sep='')
xlim = c(min(dcmpAll$x[,1])*1.10, max(dcmpAll$x[,1])*1.10)
plot( dcmpAll$x[,1:2], col=col[ metadata$Cell.Type], type='n', xlab=xlab, ylab=ylab, main="all", xlim=xlim)
text( dcmpAll$x[,1], dcmpAll$x[,2], as.character(metadata$Sample.Name), col=col[metadata$Cell.Type])
legend(0, -100, legend=levels(metadata$Cell.Type), fill=col, cex=.8)

frac = dcmpNeuron$sdev^2 / sum(dcmpNeuron$sdev^2)
xlab = paste0("PC1 (", format(100*frac[1], digits=3), "%)", sep='')
ylab = paste0("PC2 (", format(100*frac[2], digits=3), "%)", sep='')
xlim = c(min(dcmpNeuron$x[,1])*1.10, max(dcmpNeuron$x[,1])*1.10)
plot( dcmpNeuron$x[,1:2], col=col[1], type='n', xlab=xlab, ylab=ylab, main="Neuron", xlim=xlim)
text( dcmpNeuron$x[,1], dcmpNeuron$x[,2], as.character(metadata$Sample.Name[idxNeuron]), col=col[1])

frac = dcmpNPC$sdev^2 / sum(dcmpNPC$sdev^2)
xlab = paste0("PC1 (", format(100*frac[1], digits=3), "%)", sep='')
ylab = paste0("PC2 (", format(100*frac[2], digits=3), "%)", sep='')
xlim = c(min(dcmpNPC$x[,1])*1.1, max(dcmpNPC$x[,1])*1.2)
plot( dcmpNPC$x[,1:2], col=col[ metadata$Cell.Type], type='n', xlab=xlab, ylab=ylab, main="NPC", xlim=xlim)
text( dcmpNPC$x[,1], dcmpNPC$x[,2], as.character(metadata$Sample.Name[idxNPC]), col=col[2])
```

# Color PCA by CTC score
```{r PCA.color.by.CTC, fig.height=12, fig.width=4}
df = data.frame(dcmpAll$x, metadata2)
# colnames(cibersort)
figList = foreach(cellType = c("Neuron", "hiPSC", "Fibroblast")) %do% {
  ggplot(df, aes(PC1, PC2, shape=Cell.Type)) + geom_point(aes_string(colour=cellType)) + theme_bw(10) +
   theme( aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + ggtitle(paste(cellType, 'score')) +
  scale_colour_gradientn(colours=rev(setColors("red", 3)), name="score") + xlab(xlab) + ylab(ylab)
}

do.call("grid.arrange", c(figList, ncol=1))
```

# Fit linear model to test difference in CTC scores between cell types
```{r testDifference}
# x = with(df, Neuron[Cell.Type==levels(Cell.Type)[1]])
# y = with(df, Neuron[Cell.Type==levels(Cell.Type)[2]])

# with(df, wilcox.test(x,y))

ctdDifferences = foreach(cellType = colnames(cibersort), .combine=rbind) %do% {
  form = paste(cellType , '~ Cell.Type')
  coef(summary(lm(form , df)))[2,c(1,4)]
}
ctdDifferences = data.frame(ctdDifferences, Cell.Type = colnames(cibersort), stringsAsFactors=FALSE)
colnames(ctdDifferences)[2] = "pvalue"
ctdDifferences$Cell.Type[ctdDifferences$Cell.Type == "MEF"] = "Fibroblast2"
ctdDifferences$Cell.Type[ctdDifferences$Cell.Type == "Fibroblast"] = "Fibroblast1"

ctdDifferences$Cell.Type = with(ctdDifferences, factor(Cell.Type, rev(sort(Cell.Type))))

ggplot(ctdDifferences, aes(Cell.Type, Estimate)) + geom_bar(stat="identity") + coord_flip() + theme_bw() + ylab("Mean in NPCs - mean in Neurons") + xlab('')


ggplot(ctdDifferences, aes(Cell.Type, -log10(pvalue))) + geom_bar(stat="identity") + coord_flip() + theme_bw() + ylab(bquote(-log[10]~p-value)) + xlab('') + geom_hline(yintercept=-log10(0.05)) + geom_hline(yintercept=-log10(0.05/11), linetype="dashed")
```

```{r PCA.color.by.Elastin, fig.height=4, fig.width=4}
df = data.frame(dcmpOriginal$x, metadata2, gene=vobj$E['ENSG00000049540',])
frac = dcmpOriginal$sdev^2 / sum(dcmpOriginal$sdev^2)   
xlab = paste0("PC1 (", format(100*frac[1], digits=3), "%)", sep='')
ylab = paste0("PC2 (", format(100*frac[2], digits=3), "%)", sep='')


ggplot(df, aes(PC1, PC2, shape=Cell.Type)) + geom_point(aes(colour=gene)) + theme_bw(10) +
   theme( aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + ggtitle(paste('Elastin expression')) +
  scale_colour_gradientn(colours=rev(setColors("red", 3)), name=bquote(log[2]~CPM)) + xlab(xlab) + ylab(ylab)
```


# Correlation of CTC scores with principal components
```{r PCA.cell.fraction2, fig.height=7, fig.width=16}
make_plot = function( dcmp, data, nPC, n_tests, method="pearson"){

  data = data[,order(tolower(colnames(data)), decreasing=TRUE)]

  corPC = cor(dcmp$x[,1:nPC], data, method=method)

  df = melt( corPC )
  df$value[is.na(df$value)] = 0

  df$signif = c() 
  for(i in 1:nrow(df)){
    key1 = df$Var1[i]
    key2 = df$Var2[i]
    df$signif[i] = ifelse(cor.test(dcmp$x[,key1], data[,key2], method=method)$p.value < 0.05 /n_tests, "*", "")
  }

  fig = ggplot(df, aes(Var1, Var2, fill=value)) + geom_tile() + theme_bw(15) + xlab("") + ylab("Cell component") +
    scale_fill_gradient2(name = "Correlation", low='blue', mid="white", high='red', limits=c(-1,1)) + theme(aspect.ratio=ncol(corPC) / nrow(corPC),panel.grid.major = element_blank(), panel.grid.minor = element_blank())

  fig + geom_text(aes(Var1, Var2, label=signif), vjust=1) 
}

# All
data = metadata2[,colnames(metadata2) %in% colnames(cibersort)]
nPC = 2
n_tests = ncol(data) * nPC
fig1 = make_plot( dcmpAll, data, nPC, n_tests) + theme(legend.justification=c(1,1), legend.position=c(-1,.8)) + ggtitle("Combined") +
  theme(plot.title=element_text(hjust=0.5))


# NPC 
data = metadata2[idxNPC,colnames(metadata2) %in% colnames(cibersort)]
fig2 = make_plot( dcmpNPC, data, nPC, n_tests) + ggtitle("iPSC-NPC") + ylab("") + 
    theme(legend.position="none", axis.text.y=element_blank(), plot.title=element_text(hjust=0.5), 
    axis.ticks.y=element_blank())

# Neuron
data = metadata2[idxNeuron,colnames(metadata2) %in% colnames(cibersort)]
fig3 = make_plot( dcmpNeuron, data,nPC, n_tests) + ggtitle("iPSC-neuron")  + ylab("") + 
    theme(legend.position="none", axis.text.y=element_blank(), plot.title=element_text(hjust=0.5), 
    axis.ticks.y=element_blank())

grid.arrange( fig1, fig2, fig3, ncol=3)
```

```{r variancePartition, message=FALSE}
form = ~ (1|Donor) + (1| Dx) + (1|Sex) + (1|Cell.Type) + MEF + Fibroblast # +  Myocyte + Astrocytes + Neuron
vpCellFrac2 = fitExtractVarPartModel(vobj, form, metadata2)
```

```{r variancePartition.minimal, message=FALSE}
form = ~ (1|Donor) + (1| Dx) + (1|Sex) + (1|Cell.Type)
vpCellFrac_minimal = fitExtractVarPartModel(vobj, form, metadata2)
```

```{r compare.vp}
median(vpCellFrac2$Donor)
median(vpCellFrac_minimal$Donor)
mean(vpCellFrac_minimal$Donor - vpCellFrac2$Donor)
wilcox.test(vpCellFrac2$Donor, vpCellFrac_minimal$Donor, alternatve = 'greater', paired=TRUE)

```

# all cell fractions
```{r varPart.cell_fractions}
file  = "results/variancePartition/variancePartition_CTC.tsv"
df = data.frame(geneName = rownames(vpCellFrac2), vpCellFrac2)
write.table(df, file, quote=FALSE, sep="\t")

figVp = plotVarPart(sortCols(vpCellFrac2)) + theme(aspect.ratio=1/2.5)      
figVp
```

# accounting for MEF and Fibroblast increases donor contributions
```{r variancePartition_plain, message=FALSE}
form = ~ (1|Donor) + (1| Dx) + (1|Sex) + (1|Cell.Type)
vpCellFrac_plain = fitExtractVarPartModel(vobj, form, metadata2)

form = ~ (1|Donor) + (1| Dx) + (1|Sex) + (1|Cell.Type)
vpCellFrac_plain2 = fitExtractVarPartModel(exprResiduals, form, metadata2)


df = data.frame(original = with(vpCellFrac2, Donor), adjusted = with(vpCellFrac_plain, Donor ))

t.test(df$adjusted, df$original, paired=TRUE)

```


# Percent bars
```{r Supplement.vpAllbar, fig.height=8, fig.width=15, cache=FALSE, message=FALSE}
genes = c( "SNX19", 'FURIN', 'TSNARE1', 'CNTN4', 'CLCN3', 'SNAP91', 'FZD6', 'QPCT', 'PPP1R12B', "BCAN", 'PRRX1', 'SMARCE1', "UTY", 'STEAP3')

idx = match(genes, geneInfo$hgnc_symbol)

vpHGNC = sortCols(vpCellFrac2)[geneInfo$Geneid[idx],] 
rownames(vpHGNC) = geneInfo$hgnc_symbol[idx]

figPercent = plotPercentBars( vpHGNC ) + theme(aspect.ratio=1) + theme(legend.position="left")



g_legend<-function(a.gplot){ 
  tmp <- ggplot_gtable(ggplot_build(a.gplot)) 
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box") 
  legend <- tmp$grobs[[leg]] 
  return(legend)} 

leg = g_legend(figPercent)

ylab = bquote(Expression~(log[2]~CPM))

geneName = 'CNTN4'
GE = data.frame(expr = vobj$E[geneInfo$Geneid[geneInfo$geneName==geneName],], metadata2)
figStratify_CNTN4 = plotStratify( expr ~ Donor, GE, colorBy=NULL, ylab=ylab, main=geneName) + theme(aspect.ratio=1)

geneName = 'FZD6'
GE = data.frame(expr = vobj$E[geneInfo$Geneid[geneInfo$geneName==geneName],], metadata2)
# figStratify_FZD6 = plotStratify( expr ~ Dx, GE, ylab=ylab, main=geneName, sort=FALSE) + theme(aspect.ratio=1) + scale_fill_manual(values=c("blue", "red")) + scale_color_manual(values=setColors("deepskyblue2", 2), name="Diagnosis") + scale_fill_manual(values=setColors("deepskyblue2", 2), name="Diagnosis")
figStratify_FZD6 = ggplot(GE, aes(Dx, expr,fill=Cell.Type)) + geom_boxplot() + theme_bw() + ylab(bquote(Expression~(log[2]~CPM))) + ggtitle(geneName) + theme(plot.title = element_text(hjust = 0.5), aspect.ratio=1, legend.position='none') + scale_fill_manual(values = c("orange", "green2")) + xlab("Diagnosis")


geneName = 'SLITRK2'
GE = data.frame(expr = vobj$E[geneInfo$Geneid[geneInfo$geneName==geneName],], metadata2)
figStratify_SLITRK2 = plotStratify( expr ~ Dx, GE, ylab=ylab, main=geneName, sort=FALSE) + theme(aspect.ratio=1) + scale_fill_manual(values=c("blue", "red")) + scale_color_manual(values=setColors("deepskyblue2", 2), name="Diagnosis") + scale_fill_manual(values=setColors("deepskyblue2", 2), name="Diagnosis")


geneName = 'QPCT'
GE = data.frame(expr = vobj$E[geneInfo$Geneid[geneInfo$geneName==geneName],], metadata2)
# figStratify_QPCT = plotStratify( expr ~ Dx, GE, ylab=ylab, main=geneName, sort=FALSE) + theme(aspect.ratio=1) + scale_fill_manual(values=c("blue", "red")) + scale_color_manual(values=setColors("deepskyblue2", 2), name="Diagnosis") + scale_fill_manual(values=setColors("deepskyblue2", 2), name="Diagnosis")
figStratify_QPCT = ggplot(GE, aes(Dx, expr,fill=Cell.Type)) + geom_boxplot() + theme_bw() + ylab(bquote(Expression~(log[2]~CPM))) + ggtitle(geneName) + theme(plot.title = element_text(hjust = 0.5), aspect.ratio=1, legend.position='none') + scale_fill_manual(values = c("orange", "green2")) + xlab("Diagnosis") 


geneName = 'PRRX1'
GE = data.frame(expr = vobj$E[geneInfo$Geneid[geneInfo$geneName==geneName],], metadata2)

figStratify_PRRX1 = ggplot(GE, aes(Fibroblast, expr)) + geom_point() + theme_bw(14) + theme(aspect.ratio=1,
 plot.title=element_text(hjust=0.5)) + xlab("Fibroblast composition score") + ylab(ylab) + ggtitle(geneName) + 
  geom_smooth(method='lm', se=FALSE) + xlim(0, .5)

geneName = 'SMARCE1'
GE = data.frame(expr = vobj$E[geneInfo$Geneid[geneInfo$geneName==geneName],], metadata2)
figStratify_SMARCE1 = plotStratify( expr ~ Cell.Type, GE, ylab=ylab, main=geneName, sort=FALSE) + theme(aspect.ratio=1) + scale_fill_manual(values=c("orange", "green2")) + scale_color_manual(values=c("orange", "green2")) 

geneName = 'STEAP3'
GE = data.frame(expr = vobj$E[geneInfo$Geneid[geneInfo$geneName==geneName],], metadata2)
figStratify_STEAP3 = plotStratify( expr ~ Dx, GE, ylab=ylab, main=geneName, colorBy=NULL) + theme(aspect.ratio=1) 
```

```{r Fig.variancePartition, fig.width=12}
  layout = matrix(c(1,2,2,2,3,4,5,6), ncol = 4, byrow = TRUE)   
 fig = figPercent+ theme(legend.position="top") 
grid.arrange(fig, figVp, figStratify_CNTN4, figStratify_FZD6, figStratify_QPCT, figStratify_PRRX1, layout_matrix = layout, heights=c(1,1), padding = unit(0.05, "line"))       
```


```{r Fig.variancePartition1, fig.width=6}     
# grid.arrange(figStratify_SMARCE1, figStratify_STEAP3, heights=c(1), padding = unit(0.05, "line"))      
```

# VariancePartition within each cell type
```{r variancePartition_within_cellType, echo=FALSE, results="hide"}

form = ~ (1|Donor) + (1| Dx) + (1|Sex) 

vpList_by_celltype = list()
figList = list()
for( cellType in levels(metadata$Cell.Type) ){
  idx = (metadata$Cell.Type == cellType)

  vpList_by_celltype[[cellType]] = fitExtractVarPartModel( vobj[,idx], form, metadata[idx,])

  figList[[cellType]] = plotVarPart( vpList_by_celltype[[cellType]], main=cellType )
}
```

```{r variancePartition_within_cellType.plot, fig.width=14, echo=FALSE, results="hide"}
do.call("grid.arrange", c(figList, ncol=2))

for( cellType in levels(metadata$Cell.Type) ){
 
  file = paste0("results/variancePartition/variancePartition_", cellType, ".tsv")
  df = data.frame(geneName = rownames(vpList_by_celltype[[cellType]]), vpList_by_celltype[[cellType]])
  write.table(df, file, quote=FALSE, row.names=FALSE, sep="\t")
}
```

# Compare donor effect
```{r compare_donor_effect, fig.height=3.5, message=FALSE}
df = data.frame(Neuron  = vpList_by_celltype[["6_wk_FB_neuron"]]$Donor,
                NPC     = vpList_by_celltype[["NPC"]]$Donor)

resCorPearson = cor.test(df$Neuron, df$NPC, method="pearson")
resCorSpearman = cor.test(df$Neuron, df$NPC, method="spearman")

fig1 = ggplot(df, aes(NPC, Neuron)) + geom_point(size=1) + theme_bw(16) + geom_abline(color="grey")

concordance  = with(df, Neuron - NPC)

fig2 = ggplot(subset(df, Neuron + NPC > 1e-2),aes(Neuron - NPC)) + geom_histogram(aes(y = ..density..)) + theme_bw(16)

grid.arrange(fig1, fig2, ncol=2)
```

# variancePartition: include technical variables 
```{r Supplement.vpAll}
formAll = ~ (1|Donor) + (1| Dx) + (1|Sex) + (1|Cell.Type) + Fibroblast + MEF + RIN + IQ + (1|Sendai.Batch..) + (1|`hiPSC.date`) + (1|`hiPSC.technician`) + (1|`hIPSC.site`) + (1|`NPC.generation.batch`) + (1|`NPC.thaw`)

vpAll = fitExtractVarPartModel( vobj, formAll, metadata2, colinearityCutoff=.9999)
```

```{r Supplement.plot.vpAll}
vpLocal = vpAll
colnames(vpLocal) = gsub("\\.", " ", colnames(vpLocal))

file  = "results/variancePartition/variancePartition_CTC_technical_variables.tsv"
df = data.frame(geneName = rownames(vpLocal), vpLocal)
write.table(df, file, quote=FALSE, sep="\t")

plotVarPart( sortCols(vpLocal), label.angle=40 ) + theme(aspect.ratio=1/2)
```

# Correlation between covariates
```{r cor.btw.covariates_v1}
# Correlation between covariates
keys = c('RIN','IQ','Sendai.Batch..','hiPSC date','hiPSC technician','hIPSC site','NPC generation batch','NPC thaw', "SendaiCPM", "MEF", "Fibroblast", "Dx", "Sex" )
form = paste("~", paste(paste0('`',gsub(" ", "\\.", keys),'`'), collapse=" + "))
C = canCorPairs( as.formula(form), metadata2)

plotCorrMatrix( C, main="Correlation between covariates" )
```


# Differential expression between cases and controls in combined NPC + neuron dataset
```{r differential.expression_Cell.fraction}
# form = ~ Dx + Cell.Type + Neuron:Cell.Type + Myocyte_Neuron + Astrocytes_Neuron + Sex

# this give higher concordance with CMC
# form = ~ Dx + Cell.Type + Sex + MEF + Fibroblast
form = ~ Dx + Cell.Type + Sex + MEF + Fibroblast

design = model.matrix( form, metadata2)
fit = lmFit( vobj, design, block=metadata2$Donor, correlation=dupcor$consensus )
fit = eBayes(fit)

resDEall_Scz = topTable(fit, coef='DxSZ', number=1e7 , sort.by='P')
resDEall_Scz$qvalue = qvalue( resDEall_Scz$P.Value )$qvalue
resDEall_Scz$gene = geneInfo$geneName[match(rownames(resDEall_Scz), geneInfo$Geneid)]
resDEall_Scz$symbol = sub("^(ENSG.*)$", NA, resDEall_Scz$gene)
    
file = "results/differential_expression/Schizophrenia/COS_DE_combined.tsv"
df = data.frame(geneName = rownames(resDEall_Scz), resDEall_Scz)
write.table( df, file=file, row.names=FALSE, quote=FALSE, sep="\t")


# Enrichment analysis
values = resDEall_Scz$qvalue 
names(values) = resDEall_Scz$symbol
values = values[!is.na(names(values))]
resEnrich = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = .20)

file = "results/differential_expression/Schizophrenia/COS_DE_combined_enrichment.tsv"
write.table( resEnrich, file=file, row.names=TRUE, quote=FALSE, sep="\t")
```


#### Volcano plot
```{r volcano_Cell.fraction}
# Volcano plot
cutoff = 0.30
cutoff2 = 0.10
df = resDEall_Scz
col = rev(c("n.s."="grey60", "< 30%"="firebrick", "< 10%"="red"))
df$color = names(col)[3]
df$color[df$qvalue < cutoff] = names(col)[2]
df$color[df$qvalue < cutoff2] = names(col)[1]
df$color = factor(df$color, names(col))
maxlogFC = max(abs(range(df$logFC)))
figVolcano = ggplot(df, aes(logFC, -log10(P.Value), colour=color)) + geom_point() + theme_bw(15) +
  ylab(bquote(-log[10]~p-value)) + xlab(bquote(log[2]~fold~change)) + 
  scale_color_manual(values = array(col), name="FDR") + 
  geom_text_repel(data=df[df$color=="< 10%",], aes(label=gene),  min.segment.length = unit(1, "lines"), colour="black", 
    box.padding = unit(0.3, "lines"),  point.padding = unit(0.3, "lines")) + theme(legend.position=c(.9,.9), aspect.ratio=1) + 
  xlim(-maxlogFC, maxlogFC)
  
figVolcano
```

### Genes with FDR < 30%
```{r DE.genes}
kable(resDEall_Scz[resDEall_Scz$qvalue < .30,c(1:4, 7, 8)])
```

# Differential expression within each cell type
```{r DEsubsets}
metadata2$Interaction = with(metadata, gsub(" ", ".", paste( Dx, Cell.Type, sep='_')))

form = ~ Interaction - 1 + MEF + Fibroblast + Sex

design = model.matrix( form , metadata2)

cons_separate_cell_type = makeContrasts( 
    SZNPC = InteractionSZ_NPC - InteractionCT_NPC, 
    SZNeuron = InteractionSZ_6_wk_FB_neuron - InteractionCT_6_wk_FB_neuron,  
    SZinteraction = (InteractionSZ_NPC - InteractionCT_NPC) - (InteractionSZ_6_wk_FB_neuron - InteractionCT_6_wk_FB_neuron),
    levels = colnames(design))

fit = lmFit( vobj, design, block=metadata$Donor, correlation=dupcor$consensus )
fit2 <- contrasts.fit(fit, cons_separate_cell_type)
fit2 = eBayes(fit2)

resList = list()
for(key in colnames(cons_separate_cell_type)){
  res = topTable(fit2, coef=key, number=Inf, sort.by='P')
  res$qvalue = qvalue( res$P.Value )$qvalue
  res$gene = geneInfo$geneName[match(rownames(res), geneInfo$Geneid)]
  resList[[key]] = res
}

for( key in names(resList)){
  file = paste0("results/differential_expression/Schizophrenia/COS_DE_", key, ".tsv")
  write.table( data.frame(geneName = rownames(resList[[key]]), resList[[key]]), file, quote=FALSE, sep="\t", row.names=FALSE)
}
```

```{r volcano.each.celltype, fig.width=12, fig.height=6}
# Volcano plot      
figVolcanoList = foreach(cellType = names(resList)[1:2]) %do% {
   
  cutoff = 0.30 
  cutoff2 = 0.10
  df = resList[[cellType]]
  col = rev(c("n.s."="grey60", "< 30%"="firebrick", "< 10%"="red"))  
    df$color = names(col)[3]
  df$color[df$qvalue < cutoff] = names(col)[2]
  df$color[df$qvalue < cutoff2] = names(col)[1]
  df$color = factor(df$color, names(col))
  maxlogFC = max(abs(range(df$logFC)))

  qvalue_cutoff =  with(df, P.Value[which.min((qvalue - .3)^2)])
  bonferroni_cutoff = 0.30/nrow(df)

  fig = ggplot(df, aes(logFC, -log10(P.Value), colour=color)) + geom_hline(yintercept=-log10(bonferroni_cutoff), lty=3) + ylim(0, 8) +
    geom_hline(yintercept=-log10(qvalue_cutoff), lty=2, colour="firebrick") + geom_point() + theme_bw(15) +
    ylab(bquote(-log[10]~p-value)) + xlab(bquote(log[2]~fold~change)) + 
    scale_color_manual(values = array(col), name="FDR") + 
    geom_text_repel(data=df[df$color!='n.s.',], aes(label=gene),  min.segment.length = unit(1, "lines"), colour="black", 
      box.padding = unit(1, "lines"),  point.padding = unit(1, "lines")) + theme(legend.position=c(.9,.9), aspect.ratio=1) + 
    xlim(-maxlogFC, maxlogFC) + xlim(-5,5) + theme(plot.title = element_text(hjust = 0.5))

  fig = fig + annotate("text", x = 4, y = -log10(bonferroni_cutoff)*1.03, label = "30% Bonferroni", hjust=1) +
    annotate("text", x = 4, y = -log10(qvalue_cutoff)*1.05, label = "30% FDR", hjust=1) + ggtitle(cellType)

    fig    
}

do.call("grid.arrange", c(figVolcanoList, ncol=2))
```

```{r DE.each.CTC.score}
for(covariate in c('NULL',colnames(cibersort)) ){
             
  metadata2$Interaction = with(metadata, gsub(" ", ".", paste( Dx, Cell.Type, sep='_')))

  if( covariate == 'NULL' ){
    form = ~ Interaction - 1
  }else{
    form = as.formula(paste('~ Interaction - 1 + ', covariate))
  } 

  design = model.matrix( form , metadata2)

  cons_separate_cell_type = makeContrasts( 
      SZNPC = InteractionSZ_NPC - InteractionCT_NPC, 
      SZNeuron = InteractionSZ_6_wk_FB_neuron - InteractionCT_6_wk_FB_neuron,  
      SZinteraction = (InteractionSZ_NPC - InteractionCT_NPC) - (InteractionSZ_6_wk_FB_neuron - InteractionCT_6_wk_FB_neuron),
      levels = colnames(design))

  fit = lmFit( vobj, design, block=metadata$Donor, correlation=dupcor$consensus )
  fit2 <- contrasts.fit(fit, cons_separate_cell_type)
  fit2 = eBayes(fit2)

  resList_local = list()
  for(key in colnames(cons_separate_cell_type)){
    res = topTable(fit2, coef=key, number=Inf, sort.by='P')
    res$qvalue = qvalue( res$P.Value )$qvalue
    res$gene = geneInfo$geneName[match(rownames(res), geneInfo$Geneid)]
    resList_local[[key]] = res

    for( key in names(resList_local)){
      file = paste0("results/differential_expression/Schizophrenia/CTC_score/COS_DE_", key, '_', covariate,".tsv")
      write.table( data.frame(geneName = rownames(resList_local[[key]]), resList_local[[key]]), file, quote=FALSE, sep="\t", row.names=FALSE)
    }
  }
}
```

# Gene set tests
```{r gene.set.tests, cache=TRUE}

nrot = 9999
nrot_romer = 999999


# nrot = 99
# nrot_romer = 99

# Define enrichment function
enrichment_DE_analysis = function( y,index,design, contrast, block, correlation, nrot){
  
  enrichObj = list()

  # cat("romer...\n")
  enrichObj[['romer']] <- romerParallel(y=y,index=index,design=design, contrast = contrast, block=block, correlation=correlation, nrot=nrot_romer)

  # cat("roast...\n")
  enrichObj[['roast']] <- roast(y=y,index=index,design=design, contrast = contrast, block=block, correlation=correlation, nrot=nrot)
 
  return( enrichObj )
}

# sophisticated gene set enrichment 
# only retain genes with HGNC names
vobjRename = vobj
rownames(vobjRename) = geneInfo$geneName[match(rownames(vobjRename), geneInfo$Geneid)]
rownames(vobjRename) = gsub("^(ENSG.*)$", NA, rownames(vobjRename))
vobjRename = vobjRename[!is.na(rownames(vobjRename)),]

geneSetsCombined_index = ids2indices( geneSetsCombined, rownames(vobjRename))
```

```{r gene.set.tests.run}
enrichObjCellType = list()             
for( cons in colnames(cons_separate_cell_type) ){
  enrichObjCellType[[cons]] = enrichment_DE_analysis(y=vobjRename,index=geneSetsCombined_index,design=design, contrast = cons_separate_cell_type[,cons], block=metadata2$Donor, correlation=dupcor$consensus, nrot=nrot)
} 
```

NPC - roast
```{r NPC.roast, cache=FALSE}
res = enrichObjCellType[['SZNPC']][['roast']]
kable(res[1:20,]) 
```

# NPC - romer
```{r NPC.romer, cache=FALSE}
res = topRomer(enrichObjCellType[['SZNPC']][['romer']], 20, "mixed")
kable(res[1:20,]) 
```

# Neuron - roast
```{r neuron.roast, cache=FALSE}
res = enrichObjCellType[['SZNeuron']][['roast']]
kable(res[1:20,]) 
```

# Neuron - romer
```{r neuron.romer, cache=FALSE}
res = topRomer(enrichObjCellType[['SZNeuron']][['romer']], 20, "mixed")
kable(res[1:20,]) 
```

```{r save.DE, cache=FALSE}
for(key in names(enrichObjCellType)){
  for(test in names(enrichObjCellType[[key]])[1:2]){
    file = paste0('results/differential_expression/enrichment/enrichment_', gsub("SZ", '', key), '_', test, '.tsv')
    write.table(enrichObjCellType[[key]][[test]], file, quote=FALSE, sep="\t")
  }
} 
```









### Compare logFC between celltypes
```{r compare.logFC.celltypes}
resList[['SZNeuron']]$se_beta = resList[['SZNeuron']]$logFC / resList[['SZNeuron']]$t
resList[['SZNPC']]$se_beta = resList[['SZNPC']]$logFC / resList[['SZNPC']]$t

df = merge( resList[['SZNeuron']], resList[['SZNPC']], by='gene')
df = merge(df, resDEall_Scz, by="gene")

 
variables = c("gene", "logFC.x", "logFC.y", "qvalue", "se_beta.x", "se_beta.y", 't.x', 't.y')
df2 = melt( df[,variables], id.vars=variables )
df2 = df2[order(df2$qvalue, decreasing=TRUE),]

# cutoff = 0.2
# cutoff2 = 0.03
xlim = range(c( with(df2, logFC.x - se_beta.x), with(df2, logFC.x + se_beta.x)))
ylim = range(c( with(df2, logFC.y - se_beta.x), with(df2, logFC.x + se_beta.y)))


cutoff = 0.30
cutoff2 = 0.10
col = rev(c("n.s."="grey60", "< 30%"="firebrick", "< 10%"="red"))
df2$color = names(col)[3]
df2$color[df2$qvalue < cutoff] = names(col)[2]
df2$color[df2$qvalue < cutoff2] = names(col)[1]
df2$color = factor(df2$color, names(col))

ggplot(df2, aes(logFC.x, logFC.y)) + geom_errorbar(aes(ymax = logFC.y + se_beta.y, ymin=logFC.y - se_beta.y), color="grey") +
 geom_errorbarh(aes(xmax = logFC.x + se_beta.x, xmin=logFC.x - se_beta.x), color="grey") +
 geom_point(aes(color=color)) + theme_bw(15) +
  scale_color_manual(values = array(col), name="FDR") + geom_abline(color="grey") +  
  xlab( bquote(log[2]~fold~change~ln~hiPSC-neurons)) + ylab(bquote(log[2]~fold~change~ln~hiPSC-NPCs))  + 
  geom_text_repel(data=df2[df2$qvalue <cutoff2,], aes(label=gene),  min.segment.length = unit(.2, "lines"),
      box.padding = unit(.2, "lines"),  point.padding = unit(.2, "lines")) + 
  theme(legend.position=c(.1,.9), aspect.ratio=1) + xlim(range(xlim, ylim)) + ylim(range(xlim, ylim)) 

lim = range(c( with(df2, t.x), with(df2, t.y)))
ggplot(df2, aes(t.x, t.y)) + geom_point(aes(color=color)) + theme_bw(15) +
  scale_color_manual(values = array(col), name="FDR") + geom_abline(color="grey") + 
  xlab( bquote(log[2]~fold~change~ln~hiPSC-neurons)) + ylab(bquote(log[2]~fold~change~ln~hiPSC-NPCs)) +
  geom_text_repel(data=df2[df2$qvalue <cutoff2,], aes(label=gene), min.segment.length = unit(.2, "lines"), box.padding = unit(.2, "lines"),  point.padding = unit(.2, "lines")) + theme(legend.position=c(.1,.9), aspect.ratio=1) + xlim(lim) + ylim(lim)
```

## Genes with FDR < 30% within NPC or neuron separately

### Neuron
```{r DE.SZNeuron}
res = resList[['SZNeuron']]
kable(res[res$qvalue < 0.3,c(1:4, 7, 8)]) 
```

### NPC
```{r DE.SZNPC}
res = resList[['SZNPC']]
kable(res[res$qvalue < 0.3,c(1:4, 7, 8)]) 
```



```{r enrichment.prep_enrichObj_corrected}
# fancy gene set enrichment 
# vobjRename = vobj
# rownames(vobjRename) = geneInfo$geneName[match(rownames(vobjRename), geneInfo$Geneid)]
# rownames(vobjRename) = gsub("^(ENSG.*)$", NA, rownames(vobjRename))
# vobjRename = vobjRename[!is.na(rownames(vobjRename)),]

# geneSetsCombined_index = ids2indices( geneSetsCombined, rownames(vobjRename))

# enrichObj_corrected = enrichment_DE_analysis(y=vobj,index=geneSetsCombined_index,design=design, contrast = which(colnames(design) == "DxSZ"), block=metadata$Donor, correlation=dupcor$consensus, nrot=99)
```

```{r clean_names}
clean_names = function(df){
  df$`Gene Set` = sapply(strsplit(as.character(df$`Gene Set`), '_'), function(txt){
    len = length(txt)
    res = paste(txt[1:min(c(i,len))], collapse='_')

    if( len > 4 ){
      res = paste0(res, "\n", paste(txt[5:len], collapse='_'))
    }
    res
    })
  df
}
```

# Effect of correcting for Fibroblast score on case-control t-statistics in NPCs
We have demonstrated that correcting for the fibroblast score is important, especially in boosting concordance with HBCC.  So what genes are affected by this correction?  In order to test this, I compared the absolute value of the t-statstic (the test-statistic used compute the p-value) from the adjusted and unadjusted test.
```{r compare.t.NPC, fig.height=4.5, fig.width=9}
resDE_NPC_NULL = read.table(getFileLocation(synGet( 'syn9920865' )), header=TRUE, stringsAsFactors=FALSE, sep="\t")
resDE_NPC_Fibroblast = read.table(getFileLocation(synGet( 'syn9920857' )), header=TRUE, stringsAsFactors=FALSE, sep="\t")
 
resDE_NPC_NULL = resDE_NPC_NULL[order(resDE_NPC_NULL$geneName),]
resDE_NPC_Fibroblast = resDE_NPC_Fibroblast[order(resDE_NPC_Fibroblast$geneName),]

df = data.frame(gene = resDE_NPC_NULL$gene,none = abs(resDE_NPC_NULL$t), Fibroblast = abs(resDE_NPC_Fibroblast$t) )
df$difference = with(df, Fibroblast - none)
df$color = rep('remaining', nrow(df))
cutoff1 = sort(df$difference)[nrow(df) - 500]
cutoff2 = sort(df$difference)[501]
df$color[df$d > cutoff1] = "top 500"
df$color[df$d < cutoff2] = "bottom 500"
col = c("red", "blue", "black")
df$color = factor(df$color, c("top 500", "bottom 500", "remaining"))
# table(df$color)

fig1 = ggplot(df, aes(none, Fibroblast, colour=color)) + geom_point() + theme_bw(14) + scale_colour_manual(values = col, name = "Change in |t|") + geom_abline(colour="grey", lty=2) + xlab('|t| Original') + ylab('|t| after correcting for Fibroblast score') + theme(legend.position = c(.2, .8))

fig2 =  qplot(df$difference, geom="histogram", bins=50) + theme_bw(14) + xlab("Fiboblast |t| - original |t|") + geom_vline(xintercept=cutoff1, col="red", lty=2) + geom_vline(xintercept=cutoff2, col="blue", lty=2)
```

```{r compare.t.NPC.enrichment, result='asis'}
values = ifelse(df$color != "remaining", 0, 1)
names(values) = df$gene 
values = values[grep('^ENSG', names(values), invert=TRUE)]
resEnrichAll = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = 0.05)

# kable(resEnrichAll[1:3,])
kable(resEnrichAll[1:15,-1], digits=c(0, 0, 2, 2, 2, 100), row.names = FALSE)
```

## up only
```{r compare.t.NPC.enrichment.up, result='asis'}
# gene enrichment
values = ifelse(df$color == "top 500", 0, 1)
names(values) = df$gene 
values = values[grep('^ENSG', names(values), invert=TRUE)]
resEnrichUp = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = 0.05)

values = ifelse(df$color == "bottom 500", 0, 1)
names(values) = df$gene
values = values[grep('^ENSG', names(values), invert=TRUE)]
resEnrichDown = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = 0.05)

kable(resEnrichUp[1:15,-1], digits=c(0, 0, 2, 2, 2, 100), row.names = FALSE)
```

## downonly
```{r compare.t.NPC.enrichment.down, result='asis'} 
kable(resEnrichDown[1:15,-1], digits=c(0, 0, 2, 2, 2, 100), row.names = FALSE) 
```


```{r compare.t.NPC.enrichment.up.enrich, fig.height=9, fig.width=9, cache=FALSE}
res = resEnrichUp[1:10,]
res = clean_names( res )
res$`Gene Set` = factor( res$`Gene Set`,  rev(res$`Gene Set`))

fig3 = ggplot(res, aes(x=`Gene Set`, y=-log10(Pvalue))) + geom_bar(stat='identity', fill="red") + 
  coord_flip() + theme_bw(8) + ylab(bquote(Enrichment~(-log[10]~p))) + xlab('')

res = resEnrichDown[1:10,]
res = clean_names( res )
res$`Gene Set` = factor( res$`Gene Set`, rev(res$`Gene Set`))

fig4 = ggplot(res, aes(x=`Gene Set`, y=-log10(Pvalue))) + geom_bar(stat='identity', fill="blue") + 
  coord_flip() + theme_bw(8) + ylab(bquote(Enrichment~(-log[10]~p))) + xlab('')

do.call("grid.arrange", list(fig1, fig2, fig3, fig4, ncol=2))
```

# Effect of correcting for Fibroblast score on case-control t-statistics in Neurons
```{r compare.t.neuron, fig.height=4.5, fig.width=9}
resDE_neuron_NULL = read.table(getFileLocation(synGet( 'syn9920853' )), header=TRUE, stringsAsFactors=FALSE, sep="\t")
resDE_neuron_Fibroblast = read.table(getFileLocation(synGet( 'syn9920845' )), header=TRUE, stringsAsFactors=FALSE, sep="\t")

resDE_neuron_NULL = resDE_neuron_NULL[order(resDE_neuron_NULL$geneName),]
resDE_neuron_Fibroblast = resDE_neuron_Fibroblast[order(resDE_neuron_Fibroblast$geneName),] 

df = data.frame(gene = resDE_neuron_NULL$gene,none = abs(resDE_neuron_NULL$t), Fibroblast = abs(resDE_neuron_Fibroblast$t) )
df$difference = with(df, Fibroblast - none)
df$color = rep('remaining', nrow(df))
cutoff1 = sort(df$difference)[nrow(df) - 500]
cutoff2 = sort(df$difference)[501]
df$color[df$d > cutoff1] = "top 500"
df$color[df$d < cutoff2] = "bottom 500"
col = c("red", "blue", "black")
df$color = factor(df$color, c("top 500", "bottom 500", "remaining"))
# table(df$color)

fig1 = ggplot(df, aes(none, Fibroblast, colour=color)) + geom_point() + theme_bw(14) + scale_colour_manual(values = col, name = "Change in |t|") + geom_abline(colour="grey", lty=2) + xlab('|t| Original') + ylab('|t| after correcting for Fibroblast score') + theme(legend.position = c(.2, .8))

fig2 = qplot(df$difference, geom="histogram", bins=50) + theme_bw(14) + xlab("Fiboblast |t| - original |t|") + geom_vline(xintercept=cutoff1, col="red", lty=2) + geom_vline(xintercept=cutoff2, col="blue", lty=2)
```

```{r compare.t.neuron.enrichment, result='asis'}
values = ifelse(df$color != "remaining", 0, 1)
names(values) = df$gene 
values = values[grep('^ENSG', names(values), invert=TRUE)]
resEnrichAll = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = 0.05)

kable(resEnrichAll[1:15,-1], digits=c(0, 0, 2, 2, 2, 100), row.names = FALSE)
```

## up only
```{r compare.t.neuron.enrichment.up, result='asis'}
# gene enrichment
values = ifelse(df$color == "top 500", 0, 1) 
names(values) = df$gene
values = values[grep('^ENSG', names(values), invert=TRUE)]
resEnrichUp = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = 0.05)

values = ifelse(df$color == "bottom 500", 0, 1)
names(values) = df$gene
values = values[grep('^ENSG', names(values), invert=TRUE)]
resEnrichDown = enrich_mSigDB( values, list('1'=geneSetsCombined), cutoff = 0.05)

kable(resEnrichUp[1:15,-1], digits=c(0, 0, 2, 2, 2, 100), row.names = FALSE)
```

## downonly
```{r compare.t.neuron.enrichment.down, result='asis'}
kable(resEnrichDown[1:15,-1], digits=c(0, 0, 2, 2, 2, 100), row.names = FALSE)
```


```{r compare.t.neuron.enrichment.up.enrich, fig.height=9, fig.width=9}
res = resEnrichUp[1:10,] 
res = clean_names( res )
res$`Gene Set` = factor( res$`Gene Set`, rev(res$`Gene Set`))

fig3 = ggplot(res, aes(x=`Gene Set`, y=-log10(Pvalue))) + geom_bar(stat='identity', fill="red") + 
  coord_flip() + theme_bw(7) + ylab(bquote(Enrichment~(-log[10]~p))) + xlab('')


res = resEnrichDown[1:10,]
res = clean_names( res )
res$`Gene Set` = factor( res$`Gene Set`, rev(res$`Gene Set`))

fig4 = ggplot(res, aes(x=`Gene Set`, y=-log10(Pvalue))) + geom_bar(stat='identity', fill="blue") + 
  coord_flip() + theme_bw(7) + ylab(bquote(Enrichment~(-log[10]~p))) + xlab('')

do.call("grid.arrange", list(fig1, fig2, fig3, fig4, ncol=2))
```






# Effective sample size calculations
```{r ESS_figure, fig.height=7, fig.width=9, message=FALSE}
get_Ne = function(N_total, w, m, rho, scale=TRUE){
  k = N_total / (1 + (m-1)*w)

    Ne  = (m*k) / (1+rho*(m-1)) 

    if( scale ){
      Ne = Ne / N_total
    }

    return( Ne )
}

plot_Ne = function(N_total, w, m_max){

  # total cost.  Equals total number of experiments when w 1
  # N_total = 100

  # given that the first experiment from a donor has cost 1,
  # w is the cost of each successive experiment from that same donor
  # w = .1

  rho = seq(0, 1, length.out=100)
  df = data.frame(m=c(), Ne=c(), rho=c())

  m_array = c(1:m_max)
  for(m in m_array){
    
  Ne = get_Ne(N_total, w, m, rho, scale=TRUE)

    df = rbind(df, data.frame(m=as.character(m), Ne, rho))
  }

  shifter <- function(x, n = 1) {
    if (n == 0) x else c(tail(x, -n), head(x, n))
  }

  col = rev(shifter(setColors("#CD2626", m_max)))

  ggplot() + geom_line(data=df, aes(100*rho, Ne, colour=m), size=2) + theme_bw() + xlab("Donor effect (%)") + ylab("Increase in effective sample size per hiPSC line") + theme(aspect.ratio=1) + 
    ggtitle(paste("Relative cost:", 100*w, "%")) + theme(plot.title=element_text(hjust=0.5))  + 
    scale_colour_manual(values= col, name = "replicates\nper donor")
}

add_points = function(fig, rho, w, col, yval1 = 0.25){
  
  data = data.frame(m=c(1:m_max), Ne=rep(NA, m_max), rho=rho)

  for(i in 1:m_max){
    data$Ne[i] = get_Ne(100, w, i, rho/100)
  }

  a = data.frame(rho=rho, yval1=yval1)
  fig + geom_segment(data=a, aes(x = rho, y = 0, xend = rho, yend = yval1), colour = col, arrow = arrow(length = unit(0.02, "npc"))) + geom_point(data=data, aes(x=rho, y=Ne), colour=col, shape=16, size=3 )
}


vpGENESIPS = read.table(getFileLocation(synGet( 'syn9908180' )), header=TRUE, stringsAsFactors=FALSE, sep="\t")

vpMerge = merge( vpGENESIPS[,'Indiv',drop=FALSE], vpCellFrac2[,'Donor',drop=FALSE], by='row.names'  )
colnames(vpMerge) = c("gene", "hiPSC", "NPC/Neuron")
rownames(vpMerge) = vpMerge$gene
vpMerge = vpMerge[,-1]


vpcolors = c("#ffd140ff", "#00b6ebff")
xvalues = apply(vpMerge,2, median) * 100

m_max = 3

figList = list()
figList[[1]] = plot_Ne(100, 1, m_max) + ylim(0, 1) + theme(legend.justification=c(1,0), legend.position=c(.95,.70))

figList[[2]] = plot_Ne(100, .5, m_max) + ylim(0, 2.2) + theme(legend.position='none')

figList[[3]] = plot_Ne(100, .3, m_max) + ylim(0, 2.2) + theme(legend.position='none')

cost = c(1, .5, .3)
yval = c(.125, 0.25, 0.25)
i = 1
for(w in cost){
  figList[[i]] = add_points( figList[[i]], xvalues[1], w, vpcolors[1], yval[i])
  figList[[i]] = add_points( figList[[i]], xvalues[2], w, vpcolors[2], yval[i])
  i = i + 1
}

figList[[4]] = plotVarPart(vpMerge, col=vpcolors) + coord_flip() + theme_bw() + theme(legend.position="none") + scale_y_continuous(position = "right")# + ylab("Donor effect (%)")

figList[[5]] = ggplot() + theme_void()


# standardize size
gA <- ggplotGrob(figList[[1]])
gB <- ggplotGrob(figList[[4]])
maxWidth = grid::unit.pmax(gA$widths[2:5], gB$widths[2:5])
gA$widths[2:5] <- as.list(maxWidth)
gB$widths[2:5] <- as.list(maxWidth)
figList[[1]] = gA
figList[[4]] = gB

layout = matrix(c(1, 1,2, 1, 1, 3,4,4,5), nrow = 3, byrow = TRUE)

grid.arrange(figList[[1]], figList[[2]], figList[[3]], figList[[4]], figList[[5]], 
  layout_matrix = layout, heights=c(1,1, .4), padding = unit(0.05, "line"))
```

```{r CD73.CD271}
# Check Fibroblast signature, NPC and Neurons for CD73 (NT5E) and CD271 (NGFR)
geneMarker = c("NT5E", "NGFR")
ensGenes = with(geneInfo, Geneid[match(geneMarker,geneName ) ])

df2 = t(geneExprlog2RPKM[ensGenes,,drop=FALSE])
colnames(df2) = geneMarker

df2 = melt( data.frame(ID = colnames(geneExprlog2RPKM), df2, CellType=metadata$Cell.Type)) 

# ggplot(df2, aes(CellType, value)) + geom_boxplot(aes(fill=CellType)) + facet_wrap(~variable, ncol=1) + scale_fill_manual(values= c("orange", 'green2')) + theme_bw() + ylab(bquote(log[2]~RPKM)) + coord_flip()

df3 = with(df2, data.frame(Var1="A", Var2=variable, value=value, Class=CellType))
```

```{r markers_in_ref_panel}
# CD73 and CD271 in reference panel
refPanel = read.table(getFileLocation(synGet( 'syn10138996' )), header=TRUE, stringsAsFactors=FALSE, sep="\t") 

df = melt(t(refPanel[rownames(refPanel) %in% geneMarker,]))
df$Class = as.character(df$Var1)
df$Class[grep("Treutlein.MEF", df$Var1)] = "Fibroblast2"
df$Class[grep("Treutlein.Fibroblast", df$Var1)] = "Fibroblast1"
df$Class[grep("Treutlein.Myocyte", df$Var1)] = "Myocyte"
df$Class[grep("Treutlein.Neuron", df$Var1)] = "Neuron"
df$Class[grep("Choi.hiPSC", df$Var1)] = "hiPSC"
df$Class = gsub("^Zhang\\.", "", df$Class)

df$Class = factor(df$Class, rev( sort(unique(df$Class))))
df$value = log2(df$value+.1)

# ggplot(df, aes(Class, value)) + geom_boxplot() + geom_point() + theme_bw() + ylab(bquote(log[2]~RPKM)) + xlab('') + facet_wrap(~Var2) + coord_flip() + ylim(c(-3.5, 4.5))

```

# Check Fibroblast signature, NPC and Neurons for CD73 (NT5E) and CD271 (NGFR)
```{r combined_marker_plot}
df_combined = rbind(df, df3)
df_combined$fill = rep("white", nrow(df_combined))
df_combined$fill[df_combined$Class=="NPC"] = "green2"
df_combined$fill[df_combined$Class=="6_wk_FB_neuron"] = "orange"

ggplot(df_combined, aes(Class, value)) + geom_boxplot(aes(fill=fill)) + geom_point() + theme_bw() + ylab(bquote(log[2]~RPKM)) + xlab('') + facet_wrap(~Var2) + coord_flip() + ylim(c(-5, 4.5)) + scale_fill_manual(values = c("green2", "orange", "white") ) + theme(legend.position="none")
```



# SessionInfo
```{r}
sessionInfo()
```


