---
title: "hiPSC models of Schizophrenia"
subtitle: 'Public release of data and code from Hoffman, et al. (submitted)'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
date: "Run on `r Sys.Date()`"
documentclass: article
output: 
  html_document:
  toc: true
---

Data is available on Synapse.

## Summary
This includes code and plots for the final version of the paper.  Exploratory analysis and intermediate processing steps are too extensive to be included here


<!--- 
# run analysis
cd /media/sdb1/workspace/scripts/Brennand/COS/COS_public_release
rmarkdown::render("run_WGCNA.Rmd")
--->


```{r initialize, cache=FALSE, echo=FALSE, message=FALSE, results='hide'}
nthreads = 12
```

```{r load.always, cache=FALSE, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages(library(doParallel))
suppressPackageStartupMessages(library(synapseClient))

cl <- makeCluster(nthreads)
registerDoParallel(cl)

synapseLogin()

if( ! file.exists("results/coexpression/modules/") ){
  dir.create("results/coexpression/modules/", recursive=TRUE)
  dir.create("results/coexpression/enrichment/", recursive=TRUE)
}
```

```{r load.packages, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages(library(limma))
suppressPackageStartupMessages(library(edgeR))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(variancePartition))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(ape))
suppressPackageStartupMessages(library(latex2exp))
suppressPackageStartupMessages(library(readr))
suppressPackageStartupMessages(library(preprocessCore))
suppressPackageStartupMessages(library(colortools))
suppressPackageStartupMessages(library(xtable))
suppressPackageStartupMessages(library(qvalue))
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(HTSanalyzeR))
suppressPackageStartupMessages(library(GSEABase))
suppressPackageStartupMessages(library(foreach))
suppressPackageStartupMessages(library(gage))
suppressPackageStartupMessages(library(WGCNA))
allowWGCNAThreads(nthreads)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=TRUE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE,
  dev = c("png", "pdf"), 
  fig.width=7, fig.height=7)

options(markdown.HTML.stylesheet = 'css/custom.css')
```

```{r get.MSigDB, echo=FALSE, results="hide"}
# define gene set enrichment analysis code
enrich_mSigDB = function( geneScoreEntrez, mSigDB, cutoff, minGeneSetSize=20){

  res = foreach( key=names(mSigDB), .packages=c("HTSanalyzeR") ) %dopar% {

    res = multiHyperGeoTest( collectionOfGeneSets=mSigDB[[key]], universe=names(geneScoreEntrez), hits=names(geneScoreEntrez)[geneScoreEntrez <= cutoff], minGeneSetSize=minGeneSetSize, verbose=FALSE)

    res = data.frame(res, stringsAsFactors=FALSE, check.names=FALSE)

    res$Class = rep(key, nrow(res))
    res[["Gene Set"]] = rownames(res)
    rownames(res) = c()
    res[["OR"]] = res[["Observed Hits"]] / res[["Expected Hits"]]
    res
  }

  res2 = c()
  for( i in 1:length(res)){
    res2 = rbind(res2, res[[i]])
  }

  res3 = data.frame(res2[,c("Class", "Gene Set", "Gene Set Size", "Expected Hits", "Observed Hits", "OR", "Pvalue")], stringsAsFactors=FALSE, check.names=FALSE)
  
  res3 = res3[order(res3$Pvalue),]

  return( res3 )
}


read.mSigDB = function(files, standard=TRUE){

    mSigDB = list()

    for( file in files){

      if( standard ){
          key = substring(basename(file), 1,2)
        }else{
          key = sapply(strsplit(basename(file), "\\."), function(x) x[1])
        }

        mSigDB[[key]] = readList( file )
    }
    return(mSigDB)
}

# get files based on parent
geneset.files = synQuery('select * from file where parentId == "syn9908164"')

# Download files
files = sapply( geneset.files$file.id, function(file) synGet(file)@filePath)

geneSets = read.mSigDB(files, standard=FALSE )

geneSetsCombined_MSigDB = unlist( geneSets, recursive=FALSE)

i = grep("^c3", names(geneSetsCombined_MSigDB), invert=TRUE)
geneSetsCombined_MSigDB = geneSetsCombined_MSigDB[i]
i = grep("^c1", names(geneSetsCombined_MSigDB), invert=TRUE)
geneSetsCombined_MSigDB = geneSetsCombined_MSigDB[i]

# read magna genesets
geneset.files = synQuery('select * from file where parentId == "syn9908161"')

# Download files
files = sapply( geneset.files$file.id, function(file) synGet(file)@filePath)

geneSetsMagma = unlist(read.mSigDB(files, standard=FALSE), recursive=FALSE)
names(geneSetsMagma) = gsub("magma_gene_sets.", "", names(geneSetsMagma))


geneSetsCombined = c(geneSetsCombined_MSigDB, geneSetsMagma)

# remove small genesets
geneSetsCombined = geneSetsCombined[sapply(geneSetsCombined, length) > 20]
geneSetsCombined = geneSetsCombined[sapply(geneSetsCombined, length) < 5000]
# length(geneSetsCombined)
```


```{r load.data}
# get gene counts and expression cutoff
geneCounts = as.matrix(read.csv(getFileLocation(synGet( 'syn9908113' )), header=TRUE, row.names=1, check.names=FALSE))
# exonCounts = as.matrix(read.csv(getFileLocation(synGet( 'syn7113739' )), header=TRUE, row.names=1, check.names=FALSE))
isexpr = read.csv(getFileLocation(synGet( 'syn9908114' )), row.names=1, header=FALSE)[,1]
# isexpr_exon = read.csv(getFileLocation(synGet( 'syn7113733' )), row.names=1, header=FALSE)[,1]

# get gene annotations
geneInfo = read.table(getFileLocation(synGet( 'syn9909914' )), header=TRUE, stringsAsFactors=FALSE, sep="\t")

# get my metadata
info = read.csv(getFileLocation(synGet( 'syn9908174' )), header=TRUE)
info$Donor = factor(info$Donor)
idx = grep("(N|F)$", info$ID, invert=TRUE)
info$ID[idx] = gsub("-2$", "", as.character(info$ID[idx]))
colnames(geneCounts)[idx] = gsub("-2$", "R", as.character(colnames(geneCounts)[idx]))
# colnames(exonCounts)[idx] = gsub("-2$", "R", as.character(colnames(exonCounts)[idx]))

# get Brig's metdata
metadata = read.csv(getFileLocation(synGet( 'syn9908107' )), header=TRUE)
metadata$Cell.Type = factor(gsub(' ', '_',metadata$Cell.Type), c('6_wk_FB_neuron', 'NPC'))
# metadata = droplevels(metadata[-nrow(metadata),])
metadata$Donor = factor(sapply(strsplit(as.character(metadata$Sample.Name), '-'), function(x) x[1]))
metadata$Cell.Type_make.names = make.names(metadata$Cell.Type)
metadata$Exclusion = as.character(metadata$Exclusion)
# check that sample are present
# which(! (colnames(geneCounts) %in% metadata$Sample.Name))

# only keep samples present in data and metadata
idx = match(colnames(geneCounts), metadata$Sample.Name)
metadata = metadata[idx,]

# add sendai reads
sendaiReads = read.csv(getFileLocation(synGet( 'syn9908176' )), header=TRUE, stringsAsFactors=FALSE)
sendaiReads$RNA_Sample_ID = gsub("-2$", "R",sendaiReads$RNA_Sample_ID)

idx = match(metadata$Sample.Name, sendaiReads$RNA_Sample_ID)
metadata$SendaiReads = sendaiReads$SendaiCounts[idx]

# Sendai results
metadata$totalReads = colSums(geneCounts)
metadata$SendaiCPM = with(metadata, (SendaiReads) / (totalReads/1e6))

# change mislabeling
i = metadata$Sample.Name %in% c("499-X-CF", "499-X-CN")
metadata$Donor = as.character(metadata$Donor)
metadata$Donor[i] = "449"
metadata$Sex[i] = "Female"
metadata$Donor = factor(metadata$Donor)
```


```{r exclude.samples, echo=FALSE, message=FALSE}
excludeBasedOnSexGenes = c("1275-B-3F", "1275-B-3N", "1275-C-1F", "1275-C-1N", "2476-1-4F", "2476-1-4N", "2484-2aF", "2484-2aN", "3113-3-21F", "3113-3-21N", "3121-2-1F", "3121-2-1N", "3121-2-MSSM2F", "3121-2-MSSM2N", "676-1-2N")

metadata$Exclusion[metadata$Sample.Name %in% excludeBasedOnSexGenes] = 'yes' #"maybe"

# apply exclusion
idx_exclude = which(metadata$Exclusion == "yes")
```


```{r resume.processing}
metadata = metadata[-idx_exclude,]   
geneCounts = geneCounts[,-idx_exclude]
# exonCounts = exonCounts[,-idx_exclude]  
metadata$color = rainbow(nlevels(metadata$Donor))[metadata$Donor]

# check if geneCounts and metadata are sorted correctly
# table(colnames(geneCounts) == as.character(metadata$Sample.Name))
```

```{r cache=FALSE}
suppressPackageStartupMessages(library(EnsDb.Hsapiens.v70))
suppressPackageStartupMessages(library(ggbio))
txdb = EnsDb.Hsapiens.v70
data(hg19IdeogramCyto, package = "biovizBase")
```



```{r voom}
genes = DGEList(counts=geneCounts[isexpr,])
genes = calcNormFactors(genes)
design = model.matrix(~Cell.Type + Sex, metadata)

# voom / dupcor 1
#################

vobj_init = voom(genes, design, plot=FALSE )

dupcor_init = duplicateCorrelation( vobj_init, design, block=metadata$Donor)
# dupcor_init$consensus

# # voom / dupcor 2
# #################
# # re-estimate weights using correlation values

# dupcor = list(consensus = 0.56)
# dupcor_init = list(consensus = 0.56)
vobj <- voom(genes, design, block = metadata$Donor, correlation=dupcor_init$consensus)

dupcor = duplicateCorrelation( vobj, design, block=metadata$Donor)
# dupcor$consensus
```

# WGCNA
```{r WGCNA.pickSoftThreshold}

col = c("orange", "green2")

thresholdPath = foreach( cellType = levels(metadata$Cell.Type) ) %dopar% {
  idx = (cellType == metadata$Cell.Type)

  pickSoftThreshold( t(vobj$E[,idx]) )  
}
names(thresholdPath) = levels(metadata$Cell.Type)

```

# Scale-free plot
```{r WGCNA.pickSoftThreshold.plot}
with(thresholdPath[[1]]$fitIndices, plot(Power, SFT.R.sq*-sign(slope), ylab="Scale Free Topology Model Fit", xlab="Soft Threshold (power)", xlim=c(0,20), type='o', col=col[1], ylim=c(-.4, 1)))
with(thresholdPath[[2]]$fitIndices, lines(Power, SFT.R.sq*-sign(slope),type='o', col=col[2]))
legend("bottomright", legend = c("Neuron", "NPC"), fill=col)
softPower = 9
abline(v=softPower, lty=2, col="grey")
```

```{r WGCNA.TOM, cache.lazy=FALSE}

TOMlist = list()
hierTOM = list()
colorDynamicTOM= list()
for( cellType in levels(metadata$Cell.Type) ){

  cat(cellType, "\n")

  idx = (cellType == metadata$Cell.Type)

  # compute co-expression
  # ADJ = abs(cor( t(vobj$E[1:1000,idx]), use = "p"))^softPower;
  # dissADJ = 1 - ADJ
  # TOM = TOMsimilarity( ADJ )
  # dissTOM = TOMdist( ADJ )

  TOMlist[[cellType]] = 1 - TOMsimilarityFromExpr( t(vobj$E[,idx]), power=softPower, nThreads=60 )

  hierTOM[[cellType]] = hclust(as.dist(TOMlist[[cellType]]),method="average");

  colorDynamicTOM[[cellType]] = labels2colors( cutreeDynamic(hierTOM[[cellType]],method="tree") )
}

rm(TOMlist)
gc()

for( cellType in names(colorDynamicTOM) ){

  file = paste0("results/coexpression/enrichment/Coexpression_modules_", cellType, ".tsv")

  write.table( data.frame(ensGene = rownames(vobj), Module = colorDynamicTOM[[cellType]]), file, quote=FALSE, sep="\t", row.names=FALSE)
}

```

# WGCNA.tree.plots
```{r WGCNA.tree.plots.1, fig.width=10, fig.height=5}
cellType = levels(metadata$Cell.Type)[1]
  # sizeGrWindow(10,5)
  plotDendroAndColors(hierTOM[[cellType]],
    colors=data.frame(colorDynamicTOM[[cellType]]),
    dendroLabels = FALSE, marAll = c(1, 8, 3, 1),
    main = paste0(cellType, ": Gene dendrogram and module colors, TOM dissimilarity"))
```

```{r WGCNA.tree.plots.2, fig.width=10, fig.height=5}
cellType = levels(metadata$Cell.Type)[2]
  # sizeGrWindow(10,5)
  plotDendroAndColors(hierTOM[[cellType]],
    colors=data.frame(colorDynamicTOM[[cellType]]),
    dendroLabels = FALSE, marAll = c(1, 8, 3, 1),
    main = paste0(cellType, ": Gene dendrogram and module colors, TOM dissimilarity"))
```


```{r read.DE.results}
 
# Download cell type compostion scores differential expression results
# get files based on parent
files = synQuery("select * from file where parentId == 'syn9920368'")
files = files[grep("Cell_Composition_Score_DE", files$file.name),]

# Download files
resDEComponments = foreach( file = files$file.id) %do% {
  data = read.table(synGet(file)@filePath, sep="\t", header=TRUE, stringsAsFactors=FALSE)
  rownames(data) = data$geneName
  data[,-1]
}
names(resDEComponments) = gsub("Cell_Composition_Score_DE_(.*).tsv", "\\1", files$file.name)

# Download DE analysis of Scz
resDEall_Scz = read.table(synGet('syn9920404')@filePath, sep="\t", header=TRUE, stringsAsFactors=FALSE)
rownames(resDEall_Scz) = resDEall_Scz$geneName
resDEall_Scz = resDEall_Scz[,-1]

# Download DE analysis of Sendai
resDENormSendai = read.table(synGet('syn9920409')@filePath, sep="\t", header=TRUE, stringsAsFactors=FALSE)
rownames(resDENormSendai) = resDENormSendai$geneName
resDENormSendai = resDENormSendai[,-1]
```


```{r WGCNA.enrichments}

# add DE genes to geneSetsMagma
# resDEComponments
 
gsid = c(names(geneSetsMagma)[grep("CMC", names(geneSetsMagma))],
  names(geneSetsMagma)[grep("gwas", names(geneSetsMagma))],
  names(geneSetsMagma)[grep("cnv", names(geneSetsMagma))], 
  'primary_kirov_de_novo_CNV')


# Primary
geneset_a = c('SCZ-NS',
'kirov_de_novo_CNV',
'cnvgene',
'Voltage-gated_Calcium_Channel_Genes',
'sw-gwas-3-4',
'hsa-miR-137',
'PSD_(human_core)',
'ARC',
'NMDAR_network',
'PSD-95_(core)',
'mGluR5')

# secondary
geneset_b = c('AUT-NS',
'ID-NS',
'AUT-LoF',
'ID-LoF',
'CHD8',
'autppi74',
'autpp49',
'ASD',
'ID',
'FMRP-targets',
'fmrp-ascano',
'ascano_fmrp_aut_s7')

# Kiriov
geneset_c = c('Synaptic_vesicle',
'mGluR5',
'PSD-95_(core)',
'ARC+NMDAR+PSD95+mGluR5',
'ARC',
'NMDAR_network',
'Pre-synaptic_active_zone',
'Pre-synapse',
'Nucleus')

# CNV intol
geneset_e = c('HIGH',
'LOW',
'MID')

# get genes differentiall expressed with cell type

cellComponentSets = foreach( key = names(resDEComponments)) %do% {
  res = resDEComponments[[key]]

  genes = with(res, gene[qvalue < 0.05])
  genes[grep("^ENSG", genes, invert=TRUE)]
}
names(cellComponentSets) = names(resDEComponments)


genesMandatory = c( c('CMC_DE.CMC_DE_FDR_1', 'CMC_DE.CMC_DE_FDR_5', names(geneSetsMagma)[grep("CMC-mod", names(geneSetsMagma))]),
  'gws3k-8.6_Schizophrenia', 'primary_cnvgene', 'primary_kirov_de_novo_CNV', 'primary_SCZ-LoF', 'primary_ARC', 'primary_NMDAR_network', "primary_PSD_(human_core)", 'primary_PSD-95_(core)', 'primary_mGluR5', 'secondary_FMRP-targets')

geneSetsSubet = sapply(c(geneset_a, geneset_b, geneset_c, geneset_e, genesMandatory), function(id) geneSetsCombined[[id]])

geneSetsSubet[["COS_DE"]] = with(resDEall_Scz, symbol[qvalue < 0.10])
geneSetsSubet[["COS_DE"]] = geneSetsSubet[["COS_DE"]][!is.na(geneSetsSubet[["COS_DE"]] )]

geneSetsSubet = c(geneSetsSubet, cellComponentSets)
genesMandatory = c(genesMandatory, names(cellComponentSets))

geneSetsSubet[['SendaiDE']] = with(resDENormSendai, symbol[qvalue < 0.05])
geneSetsSubet[['SendaiDE']] = geneSetsSubet[['SendaiDE']][!is.na(geneSetsSubet[['SendaiDE']])]


# Enrichment for each module
enrichWGCNA = list()
for( cellType in levels(metadata$Cell.Type) ){

  cat("\r",cellType, "                     ")

  modulesArray = sort(unique(colorDynamicTOM[[cellType]]))
  enrichWGCNA[[cellType]] = foreach( module = modulesArray ) %dopar% {

    value = rep(1, length(colorDynamicTOM[[cellType]]))
    value[colorDynamicTOM[[cellType]] == module] = 0

    names(value) = geneInfo$geneName[isexpr]
    value = value[grep("^ENSG", names(value), invert=TRUE)]

    enrich_mSigDB( value, list('1'=geneSetsSubet), cutoff = .20)
  }
  names(enrichWGCNA[[cellType]]) = modulesArray
}

# Combine results
df = foreach( cellType = levels(metadata$Cell.Type), .combine=rbind ) %do% {
  foreach(module = names(enrichWGCNA[[cellType]]), .combine=rbind ) %dopar% {
    data.frame( cellType = cellType, module = module, cell_Module = paste0(cellType, '_', module),
      enrichWGCNA[[cellType]][[module]], stringsAsFactors=FALSE)    
  }
}


file = "results/coexpression/enrichment//Coexpression_enrichment.tsv"

write.table( df, file, quote=FALSE, sep="\t", row.names=FALSE)



# remove CMC -expression modules
df = df[grep("^CMC-mod", df$Gene.Set, invert=TRUE),]

modulesInclude = sapply(df$cell_Module, function(module){
    min(df$Pvalue[as.character(df$cell_Module) == module]) < 1e-4#0.05/nrow(df)
  })
table(modulesInclude)

setInclude = sapply(df$Gene.Set, function(x){
  min(df$Pvalue[as.character(df$Gene.Set) == x]) <  0.0001/nrow(df)
})

setInclude[names(setInclude) %in% genesMandatory] = TRUE
table(setInclude)


idx = (as.character(df$Gene.Set) %in% names(setInclude)[setInclude]) & 
      (as.character(df$cell_Module) %in% names(modulesInclude)[modulesInclude])

df2 = df[idx,]


colors = c("grey98", "yellow", "orange", "green", "blue", "firebrick", "red")

df2$color = rep("0 - 1", nrow(df2))
legend_label = c("0 - 1")
for( i in 1:(length(colors)-1)){  
  legend_label[i+1] = paste(i, '-', i+1)
  if(i == length(colors)){    
    legend_label[i+1] = paste('>', i+1)
  }  
  df2$color[-log10(df2$Pvalue) >= i] = legend_label[i+1]
}
names(colors) = legend_label
df2$color = factor(df2$color, legend_label)
table(df2$color)

df3 = df2[ df2$Gene.Set %in% names(cellComponentSets),]

# figCelComp = ggplot(df3, aes(Gene.Set, cell_Module, fill=pmin(-log10(Pvalue), 10))) + geom_tile(color="white") + coord_flip() +
  # scale_fill_gradient(low="grey98", high="red", name =bquote(-log[10]~P)) + xlab("Coexpression module") + theme_bw() +
  # theme(axis.text.x = element_text(angle = 90, hjust = 1),panel.grid.major = element_blank(), 
  #       panel.grid.minor = element_blank(), aspect.ratio=2) 

df3 = df2[ df2$Gene.Set %in% c('CMC_DE.CMC_DE_FDR_5', "COS_DE"),]

# figDE = ggplot(df3, aes(Gene.Set, cell_Module, fill=pmin(-log10(Pvalue), 10))) + geom_tile(color="white") + coord_flip() +
#   scale_fill_gradient(low="grey98", high="red", name =bquote(-log[10]~P)) + xlab("Coexpression module") + theme_bw() +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1),panel.grid.major = element_blank(), 
#         panel.grid.minor = element_blank(), aspect.ratio=2) 

 
# grid.arrange(figCelComp, figDE, ncol=1)
```

```{r coexpression.plot, fig.height=15, cache=FALSE}
df3 = df2[ grep("CMC-mod", df2$Gene.Set, invert=TRUE),]
 
gs_order = unique(df3$Gene.Set) 

sort_suffix = function(x, prefix){
  x[order(tolower(gsub(prefix, '', x)))]
}

gs_order2 = rev(sort_suffix(gs_order[grep("^6_wk_", gs_order)], "6_wk_FB_neuron-"))
gs_order2 = c(gs_order2, rev(sort_suffix(gs_order[grep("^NPC", gs_order)], "NPC-")))
gs_order2 = c(gs_order2, gs_order[grep("^Sendai", gs_order)])
gs_order2 = c(gs_order2, gs_order[grep("^primar", gs_order)])
gs_order2 = c(gs_order2, gs_order[grep("^second", gs_order)])
gs_order2 = c(gs_order2, gs_order[grep("Schi", gs_order)]) 
gs_order2 = c(gs_order2, gs_order[grep("DE", gs_order)][-1])

 gs_order2 = c('6_wk_FB_neuron-Oligodendrocyte_Precursor_Cell', '6_wk_FB_neuron-Neuron', '6_wk_FB_neuron-Myocyte', '6_wk_FB_neuron-MEF', '6_wk_FB_neuron-hiPSC', '6_wk_FB_neuron-Fibroblast', '6_wk_FB_neuron-Astrocytes', 'NPC-Oligodendrocyte_Precursor_Cell', 'NPC-Newly_Formed_Oligodendrocyte', 'NPC-Neuron', 'NPC-MEF', 'NPC-hiPSC', 'NPC-Fibroblast', 'NPC-Astrocytes', 'SendaiDE', 'primary_PSD_(human_core)', 'primary_ARC', 'primary_NMDAR_network', 'primary_PSD-95_(core)', 'primary_mGluR5', 'secondary_FMRP-targets', 'primary_cnvgene', 'primary_kirov_de_novo_CNV', 'primary_SCZ-LoF', 'gws3k-8.6_Schizophrenia', 'CMC_DE.CMC_DE_FDR_1', 'CMC_DE.CMC_DE_FDR_5', 'COS_DE')

keep = gs_order2[grep("Sendai|^second|^prim|Schi|DE", gs_order2)]


modulesInclude = sapply(df$cell_Module, function(module){

    obj = df3[df3$cell_Module == module,]

    any(with(obj, (Gene.Set %in% keep) & Pvalue < 0.01))
  })
table(modulesInclude)

df3 = df3[df3$cell_Module %in% unique(names(modulesInclude[modulesInclude])), ]
df3$text_dot = ifelse( (df3$Pvalue < 0.05) & (df3$Pvalue > 0.05 / nrow(df3)) , ".", '')
df3$text_star = ifelse( df3$Pvalue < 0.05 / nrow(df3), "*", '')

# make sure all sets are selected
gs_order[! gs_order %in% gs_order2]

df3$Gene.Set = factor(df3$Gene.Set, rev(gs_order2))

ratio = nlevels(df3$Gene.Set) / length(unique(df3$cell_Module))

 ggplot(df3, aes(Gene.Set, cell_Module, fill=pmin(-log10(Pvalue), 30))) + geom_tile(color="white") +
  scale_fill_gradient(low="grey98", high="red", name =bquote(-log[10]~P)) + 
  ylab("Coexpression module") + theme_bw() + xlab("Gene set") + 
    theme(axis.text.y = element_text(angle = 0, hjust = 1), axis.text.x = element_text(angle=90, hjust=1),
            panel.grid.minor = element_blank(), aspect.ratio=1/ratio) +
    geom_text(aes(label=text_dot), vjust = .1) + geom_text(aes(label=text_star), vjust = .83)
```

```{r for.paper}
# ggplot(df3, aes(Gene.Set, cell_Module, fill=pmin(-log10(Pvalue), 30))) + geom_tile(color="white") + coord_flip() +
#   scale_fill_gradient(low="grey98", high="red", name =bquote(-log[10]~P)) + 
#   ylab("Coexpression module") + theme_bw() + xlab("Gene set") + 
#     theme(axis.text.y = element_text(angle = 0, hjust = 1),
#       axis.text.x = element_blank(), panel.grid.major = element_blank(), 
#         panel.grid.minor = element_blank(), aspect.ratio=ratio) 
```








# color modules
```{r moduleColors, fig.width=15, cache=FALSE}
dfColor = with(df3, unique(data.frame(cellType, module, cell_Module)))
dfColor$cellColor = c("orange", "green2")[dfColor$cellType] 
dfColor$cellColorValue = dfColor$cellType 
 
figColor_cell = ggplot(dfColor, aes(cell_Module, cellColorValue, fill=cellColor)) + geom_tile() + 
        scale_fill_manual(values=c("green", "orange"))
# figColor_cell = figColor_cell + theme(axis.text.x = element_blank(), axis.text.y = element_blank(),panel.grid.major = element_blank(), 
  #       panel.grid.minor = element_blank(), legend.position="none") +  + coord_fixed(ratio=5)
figColor_cell = figColor_cell + theme_void() + theme(legend.position="none")



figColor_module = ggplot(dfColor, aes(cell_Module, cell_Module, fill=cell_Module)) + geom_tile() + 
            scale_fill_manual(values=as.character(dfColor$module)) 
# figColor_module = figColor_module + theme_bw(16) + theme(axis.text.x = element_blank(), axis.text.y = element_blank(), panel.grid.major = 
#                     element_blank(), panel.grid.minor = element_blank(), legend.position="none") 
figColor_module = figColor_module + theme_void() + theme(legend.position="none")


grid.arrange(figColor_cell, figColor_module, nrow=2)
```









```{r test2}
# ggplot(df2, aes(cell_Module, Gene.Set, fill=color)) + geom_tile(color="white") + coord_flip() +
#   scale_fill_manual(values= colors, name =bquote(-log[10]~P)) + xlab("Coexpression module") + theme_bw() +
#   theme(axis.text.x = element_text(angle = 30, hjust = 1),panel.grid.major = element_blank(), 
#         panel.grid.minor = element_blank()) 

  # + geom_text(data=subset(df2,Pvalue < 0.05/nrow(df2)),aes(cell_Module, Gene.Set, label="*", vjust=0.5, hjust=0.5))


# ggplot(df2, aes(cell_Module, Gene.Set, fill=-log10(Pvalue))) + geom_tile() + theme_bw() + coord_flip() +
#   scale_fill_gradient2(low = "white", high = "firebrick") +  
#   theme(axis.text.x = element_text(angle = 30, hjust = 1)) + 
#   geom_text(data=subset(df2,Pvalue < 0.05/nrow(df2)),aes(cell_Module, Gene.Set, label="*", vjust=0.5, hjust=0.5))

# DEenrich = foreach(key = names(resList)) %dopar% {
#   DE_p = resList[[key]]$logFC
#   names(DE_p) = resList[[key]]$gene
#   DE_p = DE_p[grep("^ENSG", names(DE_p), invert=TRUE)]

#   gage(DE_p, gsets = geneSetsSubet, ref = NULL, samp = NULL)
# }
# names(DEenrich) = names(resList)

# # lapply(DEenrich, function(x) head(x$greater))
```

```{r coexprDE}

EVect = list()

coexprDE = foreach( cellType = levels(metadata$Cell.Type) ) %do% {

  modulesArray = sort(unique(colorDynamicTOM[[cellType]]))

 EVect[[cellType]] = foreach( module = modulesArray, .combine=rbind ) %do% {
    i = colorDynamicTOM[[cellType]] == module
    k = (cellType == metadata$Cell.Type)

    dcmp = prcomp( t(vobj$E[i,k]) )

    PC = t(dcmp$x[,1, drop=FALSE])
    rownames(PC) = module 
    PC
  }

  idx = match(colnames(EVect[[cellType]]), metadata$Sample.Name)

  design = model.matrix(~ Dx , metadata[idx,])
 
  fit = lmFit( EVect[[cellType]], design, block=metadata$Donor[idx], correlation=dupcor$consensus )
  fit = eBayes(fit)

  topTable(fit, coef='DxSZ', number=1e7 , sort.by='P')
}
names(coexprDE) = levels(metadata$Cell.Type)
```


# Coexpression DE NPC
```{r coexprDE.NPC, results="asis", out.width=6}
kable(coexprDE[['NPC']]) 
```


# Coexpression DE NPC
```{r coexprDE.6_wk_FB_neuron, results="asis", out.width=6}
kable(coexprDE[['6_wk_FB_neuron']]) 
```

